// detritus-utils/lib/baseset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class BaseSet extends Set {
    get length() {
        return this.size;
    }
    clone() {
        return new BaseSet(this.values());
    }
    every(func) {
        return this.toArray().every(func);
    }
    filter(func) {
        const map = [];
        for (let value of this) {
            if (func(value)) {
                map.push(value);
            }
        }
        return map;
    }
    find(func) {
        for (let value of this) {
            if (func(value)) {
                return value;
            }
        }
    }
    first() {
        return this.values().next().value;
    }
    join(separator) {
        return this.toArray().join(separator);
    }
    map(func) {
        const map = [];
        for (let value of this) {
            map.push(func(value));
        }
        return map;
    }
    reduce(cb, initialValue) {
        return this.toArray().reduce(cb, initialValue);
    }
    some(func) {
        for (let value of this) {
            if (func(value)) {
                return true;
            }
        }
        return false;
    }
    sort(func) {
        return this.toArray().sort(func);
    }
    toArray() {
        return Array.from(this.values());
    }
    toJSON() {
        return this.toArray();
    }
    toString() {
        return this[Symbol.toStringTag];
    }
    get [Symbol.toStringTag]() {
        return `BaseSet (${this.size.toLocaleString()} items)`;
    }
}
exports.BaseSet = BaseSet;
