// detritus-utils/lib/timers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Interval {
    constructor() {
        this.shouldRef = true;
        this.reference = null;
    }
    get hasStarted() {
        return this.reference !== null;
    }
    hasRef() {
        return this.shouldRef;
    }
    ref() {
        this.shouldRef = true;
        if (this.reference) {
            this.reference.ref();
        }
    }
    unref() {
        this.shouldRef = false;
        if (this.reference) {
            this.reference.unref();
        }
    }
    start(milliseconds, handler) {
        this.stop();
        this.reference = setInterval(handler, milliseconds);
        if (this.shouldRef) {
            this.ref();
        }
        else {
            this.unref();
        }
    }
    stop() {
        if (this.hasStarted) {
            clearInterval(this.reference);
            this.reference = null;
        }
    }
}
exports.Interval = Interval;
class Timeout {
    constructor() {
        this.shouldRef = true;
        this.reference = null;
    }
    get hasStarted() {
        return this.reference !== null;
    }
    hasRef() {
        return this.shouldRef;
    }
    ref() {
        this.shouldRef = true;
        if (this.reference) {
            this.reference.ref();
        }
    }
    unref() {
        this.shouldRef = false;
        if (this.reference) {
            this.reference.unref();
        }
    }
    start(milliseconds, handler, override = true) {
        if (!this.hasStarted || override) {
            this.stop();
            this.reference = setTimeout(() => {
                this.reference = null;
                handler();
            }, milliseconds);
            if (this.shouldRef) {
                this.ref();
            }
            else {
                this.unref();
            }
        }
    }
    stop() {
        if (this.hasStarted) {
            clearTimeout(this.reference);
            this.reference = null;
        }
    }
}
exports.Timeout = Timeout;
function sleep(milliseconds) {
    return new Promise((resolve) => {
        setTimeout(resolve, milliseconds);
    });
}
exports.sleep = sleep;
