// detritus-client/lib/commandclient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const path = require("path");
const detritus_utils_1 = require("detritus-utils");
const client_1 = require("./client");
const clusterclient_1 = require("./clusterclient");
const constants_1 = require("./constants");
const errors_1 = require("./errors");
const utils_1 = require("./utils");
const command_1 = require("./command/command");
const context_1 = require("./command/context");
const ratelimit_1 = require("./command/ratelimit");
const collections_1 = require("./collections");
const structures_1 = require("./structures");
/**
 * Command Client, hooks onto the ShardClient to provide easier command handling
 * @category Clients
 */
class CommandClient extends detritus_utils_1.EventEmitter {
    constructor(token, options = {}) {
        super();
        this._clientListeners = {};
        this.activateOnEdits = false;
        this.ignoreMe = true;
        this.maxEditDuration = 5 * 60 * 1000;
        this.mentionsEnabled = true;
        this.ratelimits = [];
        options = Object.assign({}, options);
        if (process.env.CLUSTER_MANAGER === 'true') {
            token = process.env.CLUSTER_TOKEN;
            options.useClusterClient = true;
        }
        let client;
        if (typeof (token) === 'string') {
            if (options.useClusterClient) {
                client = new clusterclient_1.ClusterClient(token, options);
            }
            else {
                client = new client_1.ShardClient(token, options);
            }
        }
        else {
            client = token;
        }
        if (!client || !(client instanceof clusterclient_1.ClusterClient || client instanceof client_1.ShardClient)) {
            throw new Error('Token has to be a string or an instance of a client');
        }
        this.client = client;
        Object.defineProperty(this.client, 'commandClient', { value: this });
        this.activateOnEdits = !!options.activateOnEdits || this.activateOnEdits;
        this.commands = [];
        this.ignoreMe = options.ignoreMe || this.ignoreMe;
        this.maxEditDuration = +(options.maxEditDuration || this.maxEditDuration);
        this.mentionsEnabled = !!(options.mentionsEnabled || options.mentionsEnabled === undefined);
        this.onPrefixCheck = options.onPrefixCheck;
        this.prefixes = Object.freeze({
            custom: new Set(),
            mention: new Set(),
        });
        this.ran = this.client.ran;
        this.replies = new collections_1.BaseCollection({ expire: this.maxEditDuration });
        if (options.prefix !== undefined) {
            if (options.prefixes === undefined) {
                options.prefixes = [];
            }
            options.prefixes.push(options.prefix);
        }
        if (options.prefixes !== undefined) {
            options.prefixes.sort((x, y) => +(x.length < y.length));
            for (let prefix of options.prefixes) {
                prefix = prefix.trim();
                if (options.prefixSpace) {
                    prefix += ' ';
                }
                this.prefixes.custom.add(prefix);
            }
        }
        if (this.ran) {
            this.addMentionPrefixes();
        }
        if (options.ratelimit) {
            this.ratelimits.push(new ratelimit_1.CommandRatelimit(options.ratelimit));
        }
        if (options.ratelimits) {
            for (let rOptions of options.ratelimits) {
                const rType = (rOptions.type || '').toLowerCase();
                if (this.ratelimits.some((ratelimit) => ratelimit.type === rType)) {
                    throw new Error(`Ratelimit with type ${rType} already exists`);
                }
                this.ratelimits.push(new ratelimit_1.CommandRatelimit(rOptions));
            }
        }
        if (!this.prefixes.custom.size && !this.mentionsEnabled) {
            throw new Error('You must pass in prefixes or enable mentions!');
        }
        Object.defineProperties(this, {
            _clientListeners: { enumerable: false, writable: false },
            activateOnEdits: { configurable: true, writable: false },
            client: { enumerable: false, writable: false },
            commands: { writable: false },
            maxEditDuration: { configurable: true, writable: false },
            mentionsEnabled: { configurable: true, writable: false },
            onPrefixCheck: { enumerable: false },
            prefixes: { writable: false },
            prefixSpace: { configurable: true, writable: false },
            ran: { configurable: true, writable: false },
            replies: { enumerable: false, writable: false },
        });
        this._clientListeners[constants_1.ClientEvents.MESSAGE_CREATE] = null;
        this._clientListeners[constants_1.ClientEvents.MESSAGE_DELETE] = null;
        this._clientListeners[constants_1.ClientEvents.MESSAGE_UPDATE] = null;
    }
    get rest() {
        return this.client.rest;
    }
    /* Set Options */
    setActivateOnEdits(value) {
        Object.defineProperty(this, 'activateOnEdits', { value });
    }
    setMaxEditDuration(value) {
        Object.defineProperty(this, 'maxEditDuration', { value });
    }
    setMentionsEnabled(value) {
        Object.defineProperty(this, 'mentionsEnabled', { value });
    }
    setPrefixSpace(value) {
        Object.defineProperty(this, 'prefixSpace', { value });
    }
    /* Generic Command Function */
    add(options, run) {
        let command;
        if (options instanceof command_1.Command) {
            command = options;
        }
        else {
            if (typeof (options) === 'string') {
                options = { name: options, run };
            }
            else {
                if (run !== undefined) {
                    options.run = run;
                }
            }
            if (options._class === undefined) {
                command = new command_1.Command(this, options);
            }
            else {
                command = new options._class(this, options);
            }
        }
        if (typeof (command.run) !== 'function') {
            throw new Error('Command needs a run function');
        }
        for (let name of command.names) {
            if (this.commands.some((c) => c.check(name))) {
                throw new Error(`Alias/name \`${name}\` already exists.`);
            }
        }
        this.commands.push(command);
        this.commands.sort((x, y) => y.priority - x.priority);
        this.setListeners();
        return this;
    }
    addMultiple(commands = []) {
        for (let command of commands) {
            this.add(command);
        }
        return this;
    }
    async addMultipleIn(directory, isAbsolute) {
        if (!isAbsolute) {
            if (require.main) {
                // require.main.path exists but typescript doesn't let us use it..
                directory = path.join(path.dirname(require.main.filename), directory);
            }
        }
        const files = await new Promise((resolve, reject) => {
            fs.readdir(directory, (error, files) => {
                if (error) {
                    reject(error);
                }
                else {
                    resolve(files);
                }
            });
        });
        const errors = {};
        for (let file of files) {
            if (!file.endsWith('.js')) {
                continue;
            }
            const filepath = path.resolve(directory, file);
            try {
                let importedCommand = require(filepath);
                if (typeof (importedCommand) === 'function') {
                    this.add({ _file: filepath, _class: importedCommand, name: '' });
                }
                else if (importedCommand instanceof command_1.Command) {
                    Object.defineProperty(importedCommand, '_file', { value: filepath });
                    this.add(importedCommand);
                }
                else if (typeof (importedCommand) === 'object') {
                    if ('default' in importedCommand) {
                        importedCommand = importedCommand.default;
                    }
                    this.add({ ...importedCommand, _file: filepath });
                }
            }
            catch (error) {
                errors[filepath] = error;
            }
        }
        if (Object.keys(errors).length) {
            throw new errors_1.ImportedCommandsError(errors);
        }
        return this;
    }
    addMentionPrefixes() {
        let user = null;
        if (this.client instanceof clusterclient_1.ClusterClient) {
            for (let [shardId, shard] of this.client.shards) {
                if (shard.user != null) {
                    user = shard.user;
                    break;
                }
            }
        }
        else if (this.client instanceof client_1.ShardClient) {
            if (this.client.user != null) {
                user = this.client.user;
            }
        }
        if (user !== null) {
            this.prefixes.mention.add(user.mention);
            this.prefixes.mention.add(`<@!${user.id}>`);
        }
    }
    clear() {
        for (let command of this.commands) {
            if (command._file) {
                const requirePath = require.resolve(command._file);
                if (requirePath) {
                    delete require.cache[requirePath];
                }
            }
        }
        this.commands.length = 0;
        this.resetListeners();
    }
    async getAttributes(context) {
        let content = context.message.content.trim();
        let contentLower = content.toLowerCase();
        if (!content) {
            return null;
        }
        let prefix = '';
        if (this.mentionsEnabled) {
            for (let mention of this.prefixes.mention) {
                if (contentLower.startsWith(mention)) {
                    prefix = mention;
                    break;
                }
            }
        }
        if (!prefix) {
            const customPrefixes = await Promise.resolve(this.getPrefixes(context));
            for (let custom of customPrefixes) {
                if (contentLower.startsWith(custom)) {
                    prefix = custom;
                    break;
                }
            }
        }
        if (prefix) {
            content = content.substring(prefix.length).trim();
            return { content, prefix };
        }
        return null;
    }
    getCommand(attributes) {
        if (attributes.content) {
            const insensitive = attributes.content.toLowerCase();
            for (let command of this.commands) {
                const name = command.getName(insensitive);
                if (name) {
                    attributes.content = attributes.content.substring(name.length).trim();
                    return command;
                }
            }
        }
        return null;
    }
    async getPrefixes(context) {
        if (typeof (this.onPrefixCheck) === 'function') {
            const prefixes = await Promise.resolve(this.onPrefixCheck(context));
            if (prefixes instanceof Set) {
                return prefixes;
            }
            if (Array.isArray(prefixes)) {
                return new Set(prefixes);
            }
            if (typeof (prefixes) === 'string') {
                return new Set([prefixes]);
            }
            throw new Error('Invalid Prefixes Type Received');
        }
        return this.prefixes.custom;
    }
    resetListeners() {
        for (let name in this._clientListeners) {
            const listener = this._clientListeners[name];
            if (listener) {
                this.client.removeListener(name, listener);
                this._clientListeners[name] = null;
            }
        }
    }
    setListeners() {
        this.resetListeners();
        this._clientListeners[constants_1.ClientEvents.MESSAGE_CREATE] = this.handle.bind(this, constants_1.ClientEvents.MESSAGE_CREATE);
        this._clientListeners[constants_1.ClientEvents.MESSAGE_DELETE] = this.handleDelete.bind(this, constants_1.ClientEvents.MESSAGE_DELETE);
        this._clientListeners[constants_1.ClientEvents.MESSAGE_UPDATE] = this.handle.bind(this, constants_1.ClientEvents.MESSAGE_UPDATE);
        for (let name in this._clientListeners) {
            const listener = this._clientListeners[name];
            if (listener) {
                this.client.addListener(name, listener);
            }
        }
    }
    /* Kill/Run */
    kill() {
        this.client.kill();
        this.emit(constants_1.ClientEvents.KILLED);
        this.clearListeners();
    }
    async run(options = {}) {
        if (this.ran) {
            return this.client;
        }
        await this.client.run(options);
        Object.defineProperty(this, 'ran', { value: true });
        this.addMentionPrefixes();
        return this.client;
    }
    /* Handler */
    async handle(name, event) {
        const { differences, message, typing } = event;
        if (!message || (this.ignoreMe && message.fromMe)) {
            return;
        }
        if (name === constants_1.ClientEvents.MESSAGE_UPDATE) {
            if (!this.activateOnEdits) {
                return;
            }
            if (!differences || !differences.content) {
                return;
            }
        }
        const context = new context_1.Context(message, typing || null, this);
        let attributes = null;
        try {
            if (!message.fromUser) {
                throw new Error('Message is not from a user.');
            }
            if (message.isEdited) {
                const difference = message.editedAtUnix - message.timestampUnix;
                if (this.maxEditDuration < difference) {
                    throw new Error('Edit timestamp is higher than max edit duration');
                }
            }
            attributes = await this.getAttributes(context);
            if (!attributes) {
                throw new Error('Does not start with any allowed prefixes');
            }
        }
        catch (error) {
            this.emit(constants_1.ClientEvents.COMMAND_NONE, { context, error });
            return;
        }
        const command = this.getCommand(attributes);
        if (command) {
            context.command = command;
        }
        else {
            const error = new Error('Unknown Command');
            this.emit(constants_1.ClientEvents.COMMAND_NONE, { context, error });
            return;
        }
        if (!command.responseOptional && !message.canReply) {
            const error = new Error('Cannot send messages in this channel');
            this.emit(constants_1.ClientEvents.COMMAND_ERROR, { command, context, error });
            return;
        }
        if (this.ratelimits.length || command.ratelimits.length) {
            const now = Date.now();
            {
                const ratelimits = utils_1.getExceededRatelimits(this.ratelimits, message, now);
                if (ratelimits.length) {
                    const global = true;
                    const payload = { command, context, global, ratelimits, now };
                    this.emit(constants_1.ClientEvents.COMMAND_RATELIMIT, payload);
                    if (typeof (command.onRatelimit) === 'function') {
                        try {
                            await Promise.resolve(command.onRatelimit(context, ratelimits, { global, now }));
                        }
                        catch (error) {
                            // do something with this error?
                        }
                    }
                    return;
                }
            }
            {
                const ratelimits = utils_1.getExceededRatelimits(command.ratelimits, message, now);
                if (ratelimits.length) {
                    const global = false;
                    const payload = { command, context, global, ratelimits, now };
                    this.emit(constants_1.ClientEvents.COMMAND_RATELIMIT, payload);
                    if (typeof (command.onRatelimit) === 'function') {
                        try {
                            await Promise.resolve(command.onRatelimit(context, ratelimits, { global, now }));
                        }
                        catch (error) {
                            // do something with this error?
                        }
                    }
                    return;
                }
            }
        }
        if (command.disableDm && context.inDm) {
            const error = new Error('Command with DMs disabled used in DM');
            if (command.disableDmReply) {
                this.emit(constants_1.ClientEvents.COMMAND_ERROR, { command, context, error });
            }
            else {
                try {
                    const reply = await message.reply(`Cannot use \`${command.name}\` in DMs.`);
                    if (this.maxEditDuration) {
                        this.replies.set(message.id, reply);
                    }
                    this.emit(constants_1.ClientEvents.COMMAND_ERROR, { command, context, error, reply });
                }
                catch (e) {
                    this.emit(constants_1.ClientEvents.COMMAND_ERROR, { command, context, error, extra: e });
                }
            }
            return;
        }
        if (typeof (command.onBefore) === 'function') {
            try {
                const shouldContinue = await Promise.resolve(command.onBefore(context));
                if (!shouldContinue) {
                    if (typeof (command.onCancel) === 'function') {
                        const reply = await Promise.resolve(command.onCancel(context));
                        if (this.maxEditDuration && reply instanceof structures_1.Message) {
                            this.replies.set(message.id, reply);
                        }
                    }
                    return;
                }
            }
            catch (error) {
                this.emit(constants_1.ClientEvents.COMMAND_ERROR, { command, context, error });
                return;
            }
        }
        const prefix = context.prefix = attributes.prefix;
        const { errors, parsed: args } = await command.getArgs(attributes, context);
        if (Object.keys(errors).length) {
            if (typeof (command.onTypeError) === 'function') {
                const reply = await Promise.resolve(command.onTypeError(context, args, errors));
                if (this.maxEditDuration && reply instanceof structures_1.Message) {
                    this.replies.set(message.id, reply);
                }
            }
            const error = new Error('Command errored out while converting args');
            this.emit(constants_1.ClientEvents.COMMAND_ERROR, { command, context, error, extra: errors });
            return;
        }
        try {
            if (typeof (command.onBeforeRun) === 'function') {
                const shouldRun = await Promise.resolve(command.onBeforeRun(context, args));
                if (!shouldRun) {
                    if (typeof (command.onCancelRun) === 'function') {
                        const reply = await Promise.resolve(command.onCancelRun(context, args));
                        if (this.maxEditDuration && reply instanceof structures_1.Message) {
                            this.replies.set(message.id, reply);
                        }
                    }
                    return;
                }
            }
            try {
                if (typeof (command.run) === 'function') {
                    const reply = await Promise.resolve(command.run(context, args));
                    if (this.maxEditDuration && reply instanceof structures_1.Message) {
                        this.replies.set(message.id, reply);
                    }
                }
                this.emit(constants_1.ClientEvents.COMMAND_RAN, { args, command, context, prefix });
                if (typeof (command.onSuccess) === 'function') {
                    await Promise.resolve(command.onSuccess(context, args));
                }
            }
            catch (error) {
                this.emit(constants_1.ClientEvents.COMMAND_RUN_ERROR, { args, command, context, error, prefix });
                if (typeof (command.onRunError) === 'function') {
                    await Promise.resolve(command.onRunError(context, args, error));
                }
            }
        }
        catch (error) {
            if (typeof (command.onError) === 'function') {
                await Promise.resolve(command.onError(context, args, error));
            }
            this.emit(constants_1.ClientEvents.COMMAND_FAIL, { args, command, context, error, prefix });
        }
    }
    async handleDelete(name, payload) {
        this.replies.delete(payload.raw.id);
    }
    on(event, listener) {
        super.on(event, listener);
        return this;
    }
}
exports.CommandClient = CommandClient;
