// detritus-client/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const detritus_client_rest_1 = require("detritus-client-rest");
const detritus_client_socket_1 = require("detritus-client-socket");
const detritus_utils_1 = require("detritus-utils");
const { AuthTypes, DiscordAbortCodes, HTTPMethods, } = detritus_client_rest_1.Constants;
exports.AuthTypes = AuthTypes;
exports.DiscordAbortCodes = DiscordAbortCodes;
exports.HTTPMethods = HTTPMethods;
const { CompressTypes, EncodingTypes, GatewayActivityActionTypes, GatewayActivityFlags, GatewayActivityTypes, GatewayDispatchEvents, GatewayOpCodes, GatewayPresenceStatuses, MediaOpCodes, MediaSpeakingFlags, SocketCloseCodes, SocketGatewayCloseCodes, SocketMediaCloseCodes, SocketStates, SocketEvents, DEFAULT_SHARD_LAUNCH_DELAY, } = detritus_client_socket_1.Constants;
exports.CompressTypes = CompressTypes;
exports.EncodingTypes = EncodingTypes;
exports.ActivityActionTypes = GatewayActivityActionTypes;
exports.ActivityFlags = GatewayActivityFlags;
exports.ActivityTypes = GatewayActivityTypes;
exports.GatewayDispatchEvents = GatewayDispatchEvents;
exports.GatewayOpCodes = GatewayOpCodes;
exports.PresenceStatuses = GatewayPresenceStatuses;
exports.MediaOpCodes = MediaOpCodes;
exports.SpeakingFlags = MediaSpeakingFlags;
exports.SocketCloseCodes = SocketCloseCodes;
exports.SocketGatewayCloseCodes = SocketGatewayCloseCodes;
exports.SocketMediaCloseCodes = SocketMediaCloseCodes;
exports.SocketStates = SocketStates;
exports.DEFAULT_SHARD_LAUNCH_DELAY = DEFAULT_SHARD_LAUNCH_DELAY;
const { DISCORD_SNOWFLAKE_EPOCH, DISCORD_TOKEN_EPOCH, } = detritus_utils_1.Constants;
exports.DISCORD_SNOWFLAKE_EPOCH = DISCORD_SNOWFLAKE_EPOCH;
exports.DISCORD_TOKEN_EPOCH = DISCORD_TOKEN_EPOCH;
exports.Package = Object.freeze({
    URL: 'https://github.com/detritusjs/client',
    VERSION: '0.8.10',
});
exports.DEFAULT_MAX_MEMBERS = 250000;
exports.DEFAULT_MAX_PRESENCES = 5000;
exports.LOCAL_GUILD_ID = '@me';
exports.MAX_ATTACHMENT_SIZE = 8 * 1024 * 1024;
exports.MAX_ATTACHMENT_SIZE_PREMIUM = 50 * 1024 * 1024;
exports.MAX_BITRATE = 96000;
exports.MAX_EMOJI_SIZE = 256000;
exports.MAX_EMOJI_SLOTS = 50;
exports.MAX_EMOJI_SLOTS_MORE = 200;
exports.MIN_BITRATE = 8000;
exports.SPOILER_ATTACHMENT_PREFIX = 'SPOILER_';
exports.ApplicationNewsFlags = Object.freeze({
    PATCH_NOTES: 1 << 1,
    PROMOTION: 1 << 2,
});
exports.Oauth2AssetTypes = Object.freeze({
    SMALL: 1,
    LARGE: 2,
});
exports.AuditLogActions = Object.freeze({
    GUILD_UPDATE: 1,
    CHANNEL_CREATE: 10,
    CHANNEL_UPDATE: 11,
    CHANNEL_DELETE: 12,
    CHANNEL_OVERWRITE_CREATE: 13,
    CHANNEL_OVERWRITE_UPDATE: 14,
    CHANNEL_OVERWRITE_DELETE: 15,
    MEMBER_KICK: 20,
    MEMBER_PRUNE: 21,
    MEMBER_BAN_ADD: 22,
    MEMBER_BAN_REMOVE: 23,
    MEMBER_UPDATE: 24,
    MEMBER_ROLE_UPDATE: 25,
    ROLE_CREATE: 30,
    ROLE_UPDATE: 31,
    ROLE_DELETE: 32,
    INVITE_CREATE: 40,
    INVITE_UPDATE: 41,
    INVITE_DELETE: 42,
    WEBHOOK_CREATE: 50,
    WEBHOOK_UPDATE: 51,
    WEBHOOK_DELETE: 52,
    EMOJI_CREATE: 60,
    EMOJI_UPDATE: 61,
    EMOJI_DELETE: 62,
    MESSAGE_DELETE: 72,
});
exports.AuditLogActionTypes = detritus_utils_1.Tools.normalize({
    ALL: null,
    CREATE: null,
    UPDATE: null,
    DELETE: null,
});
exports.AuditLogSubtargetTypes = Object.freeze({
    USER: 'member',
    ROLE: 'role',
});
exports.AuditLogTargetTypes = detritus_utils_1.Tools.normalize({
    ALL: null,
    GUILD: null,
    CHANNEL: null,
    USER: null,
    ROLE: null,
    INVITE: null,
    WEBHOOK: null,
    EMOJI: null,
});
exports.AuditLogChangeKeys = Object.freeze({
    AFK_CHANNEL_ID: 'afk_channel_id',
    AFK_TIMEOUT: 'afk_timeout',
    APPLICATION_ID: 'application_id',
    AVATAR_HASH: 'avatar_hash',
    BANNER_HASH: 'banner_hash',
    BITRATE: 'bitrate',
    CHANNEL_ID: 'channel_id',
    CODE: 'code',
    COLOR: 'color',
    DEAF: 'deaf',
    DEFAULT_MESSAGE_NOTIFICATIONS: 'default_message_notifications',
    DESCRIPTION: 'description',
    EXPLICIT_CONTENT_FILTER: 'explicit_content_filter',
    HOIST: 'hoist',
    ICON_HASH: 'icon_hash',
    ID: 'id',
    INVITER_ID: 'inviter_id',
    MAX_AGE: 'max_age',
    MAX_USES: 'max_uses',
    MENTIONABLE: 'mentionable',
    MFA_LEVEL: 'mfa_level',
    MUTE: 'mute',
    NAME: 'name',
    NICK: 'nick',
    NSFW: 'nsfw',
    OWNER_ID: 'owner_id',
    PERMISSION_OVERWRITES: 'permission_overwrites',
    PERMISSIONS: 'permissions',
    PERMISSIONS_DENIED: 'deny',
    PERMISSIONS_GRANTED: 'allow',
    POSITION: 'position',
    PRUNE_DELETE_DAYS: 'prune_delete_days',
    RATE_LIMIT_PER_USER: 'rate_limit_per_user',
    REASON: 'reason',
    REGION: 'region',
    ROLES_ADD: '$add',
    ROLES_REMOVE: '$remove',
    SPLASH_HASH: 'splash_hash',
    SYSTEM_CHANNEL_ID: 'system_channel_id',
    WIDGET_CHANNEL_ID: 'widget_channel_id',
    WIDGET_ENABLED: 'widget_enabled',
    VANITY_URL_CODE: 'vanity_url_code',
    VERIFICATION_LEVEL: 'verification_level',
    TEMPORARY: 'temporary',
    TOPIC: 'topic',
    TYPE: 'type',
    USES: 'uses',
});
exports.ChannelTypes = Object.freeze({
    BASE: -1,
    GUILD_TEXT: 0,
    DM: 1,
    GUILD_VOICE: 2,
    GROUP_DM: 3,
    GUILD_CATEGORY: 4,
    GUILD_NEWS: 5,
    GUILD_STORE: 6,
    GUILD_LFG_LISTINGS: 7,
});
exports.ClientEvents = Object.freeze({
    ACTIVITY_JOIN_INVITE: 'activityJoinInvite',
    ACTIVITY_JOIN_REQUEST: 'activityJoinRequest',
    ACTIVITY_START: 'activityStart',
    BRAINTREE_POPUP_BRIDGE_CALLBACK: 'braintreePopupBridgeCallback',
    CALL_CREATE: 'callCreate',
    CALL_DELETE: 'callDelete',
    CALL_UPDATE: 'callUpdate',
    CHANNEL_CREATE: 'channelCreate',
    CHANNEL_DELETE: 'channelDelete',
    CHANNEL_UPDATE: 'channelUpdate',
    CHANNEL_PINS_ACK: 'channelPinsAck',
    CHANNEL_PINS_UPDATE: 'channelPinsUpdate',
    CHANNEL_RECIPIENT_ADD: 'channelRecipientAdd',
    CHANNEL_RECIPIENT_REMOVE: 'channelRecipientRemove',
    ENTITLEMENT_CREATE: 'entitlementCreate',
    ENTITLEMENT_DELETE: 'entitlementDelete',
    ENTITLEMENT_UPDATE: 'entitlementUpdate',
    FRIEND_SUGGESTION_CREATE: 'friendSuggestionCreate',
    FRIEND_SUGGESTION_DELETE: 'friendSuggestionDelete',
    GIFT_CODE_UPDATE: 'giftCodeUpdate',
    GUILD_BAN_ADD: 'guildBanAdd',
    GUILD_BAN_REMOVE: 'guildBanRemove',
    GUILD_CREATE: 'guildCreate',
    GUILD_DELETE: 'guildDelete',
    GUILD_UPDATE: 'guildUpdate',
    GUILD_EMOJIS_UPDATE: 'guildEmojisUpdate',
    GUILD_INTEGRATIONS_UPDATE: 'guildIntegrationsUpdate',
    GUILD_MEMBER_ADD: 'guildMemberAdd',
    GUILD_MEMBER_LIST_UPDATE: 'guildMemberListUpdate',
    GUILD_MEMBER_REMOVE: 'guildMemberRemove',
    GUILD_MEMBER_UPDATE: 'guildMemberUpdate',
    GUILD_MEMBERS_CHUNK: 'guildMembersChunk',
    GUILD_ROLE_CREATE: 'guildRoleCreate',
    GUILD_ROLE_DELETE: 'guildRoleDelete',
    GUILD_ROLE_UPDATE: 'guildRoleUpdate',
    LIBRARY_APPLICATION_UPDATE: 'libraryApplicationUpdate',
    LOBBY_CREATE: 'lobbyCreate',
    LOBBY_DELETE: 'lobbyDelete',
    LOBBY_UPDATE: 'lobbyUpdate',
    LOBBY_MEMBER_CONNECT: 'lobbyMemberConnect',
    LOBBY_MEMBER_DISCONNECT: 'lobbyMemberDisconnect',
    LOBBY_MEMBER_UPDATE: 'lobbyMemberUpdate',
    LOBBY_MESSAGE: 'lobbyMessage',
    LOBBY_VOICE_SERVER_UPDATE: 'lobbyVoiceServerUpdate',
    LOBBY_VOICE_STATE_UPDATE: 'lobbyVoiceStateUpdate',
    MESSAGE_ACK: 'messageAck',
    MESSAGE_CREATE: 'messageCreate',
    MESSAGE_DELETE: 'messageDelete',
    MESSAGE_DELETE_BULK: 'messageDeleteBulk',
    MESSAGE_REACTION_ADD: 'messageReactionAdd',
    MESSAGE_REACTION_REMOVE: 'messageReactionRemove',
    MESSAGE_REACTION_REMOVE_ALL: 'messageReactionRemoveAll',
    MESSAGE_UPDATE: 'messageUpdate',
    OAUTH2_TOKEN_REMOVE: 'oauth2TokenRemove',
    PRESENCES_REPLACE: 'presencesReplace',
    PRESENCE_UPDATE: 'presenceUpdate',
    RECENT_MENTION_DELETE: 'recentMentionDelete',
    RELATIONSHIP_ADD: 'relationshipAdd',
    RELATIONSHIP_REMOVE: 'relationshipRemove',
    SESSIONS_REPLACE: 'sessionsReplace',
    STREAM_CREATE: 'streamCreate',
    STREAM_DELETE: 'streamDelete',
    STREAM_SERVER_UPDATE: 'streamServerUpdate',
    STREAM_UPDATE: 'streamUpdate',
    TYPING_START: 'typingStart',
    TYPING_STOP: 'typingStop',
    USER_ACHIEVEMENT_UPDATE: 'userAchievementUpdate',
    USER_CONNECTIONS_UPDATE: 'userConnectionsUpdate',
    USER_FEED_SETTINGS_UPDATE: 'userFeedSettingsUpdate',
    USER_GUILD_SETTINGS_UPDATE: 'userGuildSettingsUpdate',
    USER_NOTE_UPDATE: 'userNoteUpdate',
    USER_PAYMENT_SOURCES_UPDATE: 'userPaymentSourcesUpdate',
    USER_PAYMENTS_UPDATE: 'userPaymentsUpdate',
    USER_REQUIRED_ACTION_UPDATE: 'userRequiredActionUpdate',
    USER_SETTINGS_UPDATE: 'userSettingsUpdate',
    USER_SUBSCRIPTIONS_UPDATE: 'userSubscriptionsUpdate',
    USER_UPDATE: 'userUpdate',
    VOICE_SERVER_UPDATE: 'voiceServerUpdate',
    VOICE_STATE_UPDATE: 'voiceStateUpdate',
    WEBHOOKS_UPDATE: 'webhooksUpdate',
    CLUSTER_PROCESS: 'clusterProcess',
    COMMAND_ERROR: 'commandError',
    COMMAND_FAIL: 'commandFail',
    COMMAND_NONE: 'commandNone',
    COMMAND_RAN: 'commandRan',
    COMMAND_RATELIMIT: 'commandRatelimit',
    COMMAND_RUN_ERROR: 'commandRunError',
    GATEWAY_READY: 'gatewayReady',
    GATEWAY_RESUMED: 'gatewayResumed',
    KILLED: 'killed',
    RAW: 'raw',
    READY: 'ready',
    REST_REQUEST: 'restRequest',
    REST_RESPONSE: 'restResponse',
    SHARD: 'shard',
    UNKNOWN: 'unknown',
    WARN: 'warn',
});
exports.ClusterIPCOpCodes = Object.freeze({
    READY: 0,
    CLOSE: 1,
    RECONNECTING: 2,
    RESPAWN_ALL: 3,
    EVAL: 4,
});
exports.Colors = Object.freeze({
    BLURPLE: 7506394,
});
exports.CommandArgumentTypes = Object.freeze({
    BOOL: 'bool',
    FLOAT: 'float',
    NUMBER: 'number',
    STRING: 'string',
});
exports.CommandErrors = Object.freeze({});
exports.CommandRatelimitTypes = Object.freeze({
    CHANNEL: 'channel',
    GUILD: 'guild',
    USER: 'user',
});
exports.DiscordOpusFormat = Object.freeze({
    CHANNELS: 2,
    SAMPLE_RATE: 48000,
});
exports.DiscordRegexNames = detritus_utils_1.Tools.normalize({
    EMOJI: null,
    MENTION_CHANNEL: null,
    MENTION_ROLE: null,
    MENTION_USER: null,
    TEXT_BOLD: null,
    TEXT_CODEBLOCK: null,
    TEXT_CODESTRING: null,
    TEXT_ITALICS: null,
    TEXT_SNOWFLAKE: null,
    TEXT_SPOILER: null,
    TEXT_STRIKE: null,
    TEXT_UNDERLINE: null,
    TEXT_URL: null,
});
exports.DiscordRegex = Object.freeze({
    [exports.DiscordRegexNames.EMOJI]: /<a?:(\w+):(\d+)>/g,
    [exports.DiscordRegexNames.MENTION_CHANNEL]: /<#(\d+)>/g,
    [exports.DiscordRegexNames.MENTION_ROLE]: /<@&(\d+)>/g,
    [exports.DiscordRegexNames.MENTION_USER]: /<@(!?)(\d+)>/g,
    [exports.DiscordRegexNames.TEXT_BOLD]: /\*\*([\s\S]+?)\*\*/g,
    [exports.DiscordRegexNames.TEXT_CODEBLOCK]: /```(([a-z0-9-]+?)\n+)?\n*([^]+?)\n*```/gi,
    [exports.DiscordRegexNames.TEXT_CODESTRING]: /`([\s\S]+?)`/g,
    [exports.DiscordRegexNames.TEXT_ITALICS]: /_([\s\S]+?)_|\*([\s\S]+?)\*/g,
    [exports.DiscordRegexNames.TEXT_SNOWFLAKE]: /(\d+)/g,
    [exports.DiscordRegexNames.TEXT_SPOILER]: /\|\|([\s\S]+?)\|\|/g,
    [exports.DiscordRegexNames.TEXT_STRIKE]: /~~([\s\S]+?)~~(?!_)/g,
    [exports.DiscordRegexNames.TEXT_UNDERLINE]: /__([\s\S]+?)__/g,
    [exports.DiscordRegexNames.TEXT_URL]: /((?:https?):\/\/[^\s<]+[^<.,:;"'\]\s])/g,
});
exports.Distributors = Object.freeze({
    BATTLENET: 'battlenet',
    DISCORD: 'discord',
    EPIC: 'epic',
    GOG: 'gog',
    ORIGIN: 'origin',
    STEAM: 'steam',
    TWITCH: 'twitch',
    UPLAY: 'uplay',
});
exports.DistributorNames = Object.freeze({
    [exports.Distributors.BATTLENET]: 'Battle.net',
    [exports.Distributors.DISCORD]: 'Discord',
    [exports.Distributors.EPIC]: 'Epic',
    [exports.Distributors.GOG]: 'GOG',
    [exports.Distributors.ORIGIN]: 'Origin',
    [exports.Distributors.STEAM]: 'Steam',
    [exports.Distributors.TWITCH]: 'Twitch',
    [exports.Distributors.UPLAY]: 'Uplay',
});
// twitch shut down
exports.DistributorUrls = detritus_utils_1.Tools.URIEncodeWrap({
    [exports.Distributors.BATTLENET]: (skuId) => `https://shop.battle.net/family/${skuId}`,
    [exports.Distributors.DISCORD]: (skuId, slug) => detritus_client_rest_1.Endpoints.Routes.URL + `/store/skus/${skuId}` + ((slug) ? `/${slug}` : ''),
    [exports.Distributors.EPIC]: (skuId) => `https://epicgames.com/store/product/${skuId}`,
    [exports.Distributors.GOG]: (skuId) => `https://gog.com/game/${skuId}`,
    [exports.Distributors.ORIGIN]: (skuId) => `https://origin.com/search?searchString=${skuId}`,
    [exports.Distributors.STEAM]: (skuId) => `https://store.steampowered.com/app/${skuId}`,
    [exports.Distributors.UPLAY]: (skuId) => `https://store.ubi.com/search/?q=${skuId}`,
});
exports.ExplicitContentFilterTypes = Object.freeze({
    DISABLED: 0,
    NON_FRIENDS: 1,
    FRIENDS_AND_NON_FRIENDS: 2,
});
exports.GuildExplicitContentFilterTypes = Object.freeze({
    DISABLED: 0,
    MEMBERS_WITHOUT_ROLES: 1,
    ALL_MEMBERS: 2,
});
exports.GuildFeatures = detritus_utils_1.Tools.normalize({
    ANIMATED_ICON: null,
    BANNER: null,
    COMMERCE: null,
    DISCOVERABLE: null,
    FEATURABLE: null,
    INVITE_SPLASH: null,
    LURKABLE: null,
    MORE_EMOJI: null,
    NEWS: null,
    PARTNERED: null,
    PUBLIC: null,
    VANITY_URL: null,
    VERIFIED: null,
    VIP_REGIONS: null,
});
exports.GuildNotificationSettings = Object.freeze({
    ALL: 0,
    MENTIONS: 1,
});
exports.ImageFormats = Object.freeze({
    GIF: 'gif',
    JPEG: 'jpeg',
    JPG: 'jpg',
    PNG: 'png',
    WEBP: 'webp',
});
exports.InviteTargetUserTypes = Object.freeze({
    STREAM: 1,
});
exports.LobbyErrors = Object.freeze({
    NO_ERROR: 0,
    UNKNOWN_ERROR: 1,
    SERVICE_UNAVAILABLE: 2,
    NOT_FOUND: 3,
    INVALID_SECRET: 4,
    FULL: 5,
    LOBBY_LIMIT_REACHED: 6,
    ALREADY_CONNECTING: 7,
});
exports.LobbyTypes = Object.freeze({
    PRIVATE: 1,
    PUBLIC: 2,
});
exports.Locales = Object.freeze({
    BULGARIAN: 'bg',
    CHINESE: 'zh-CN',
    CHINESE_TAIWAN: 'zh-TW',
    CROATIAN: 'hr',
    CZECH: 'cs',
    DANISH: 'da',
    DUTCH: 'nl',
    ENGLISH_GB: 'en-GB',
    ENGLISH_US: 'en-US',
    FINNISH: 'fi',
    FRENCH: 'fr',
    GERMAN: 'de',
    GREEK: 'el',
    HUNGARIAN: 'hu',
    ITALIAN: 'it',
    JAPANESE: 'ja',
    KOREAN: 'ko',
    LITHUANIAN: 'lt',
    NORWEGIAN: 'no',
    POLISH: 'pl',
    PORTUGUESE: 'pt-PT',
    PORTUGUESE_BRAZILIAN: 'pt-BR',
    ROMANIAN: 'ro',
    RUSSIAN: 'ru',
    SLOVAK: 'sk',
    SPANISH: 'es-ES',
    SWEDISH: 'sv-SE',
    THAI: 'th',
    TURKISH: 'tr',
    UKRAINIAN: 'uk',
    VIETNAMESE: 'vi',
});
exports.LocalesText = Object.freeze({
    [exports.Locales.BULGARIAN]: 'Bulgarian',
    [exports.Locales.CHINESE]: 'Chinese, China',
    [exports.Locales.CHINESE_TAIWAN]: 'Chinese, Taiwan',
    [exports.Locales.CROATIAN]: 'Croatian',
    [exports.Locales.CZECH]: 'Czech',
    [exports.Locales.DANISH]: 'Danish',
    [exports.Locales.DUTCH]: 'Dutch',
    [exports.Locales.ENGLISH_GB]: 'English, UK',
    [exports.Locales.ENGLISH_US]: 'English, US',
    [exports.Locales.FINNISH]: 'Finnish',
    [exports.Locales.FRENCH]: 'French',
    [exports.Locales.GERMAN]: 'German',
    [exports.Locales.GREEK]: 'Greek',
    [exports.Locales.HUNGARIAN]: 'Hungarian',
    [exports.Locales.ITALIAN]: 'Italian',
    [exports.Locales.JAPANESE]: 'Japanese',
    [exports.Locales.KOREAN]: 'Korean',
    [exports.Locales.LITHUANIAN]: 'Lithuanian',
    [exports.Locales.NORWEGIAN]: 'Norwegian',
    [exports.Locales.POLISH]: 'Polish',
    [exports.Locales.PORTUGUESE]: 'Portuguese',
    [exports.Locales.PORTUGUESE_BRAZILIAN]: 'Portuguese, Brazilian',
    [exports.Locales.ROMANIAN]: 'Romanian',
    [exports.Locales.RUSSIAN]: 'Russian',
    [exports.Locales.SLOVAK]: 'Slovak',
    [exports.Locales.SPANISH]: 'Spanish',
    [exports.Locales.SWEDISH]: 'Swedish',
    [exports.Locales.THAI]: 'Thai',
    [exports.Locales.TURKISH]: 'Turkish',
    [exports.Locales.UKRAINIAN]: 'Ukrainian',
    [exports.Locales.VIETNAMESE]: 'Vietnamese',
});
exports.MessageEmbedTypes = Object.freeze({
    APPLICATION_NEWS: 'application_news',
    ARTICLE: 'article',
    GIFV: 'gifv',
    IMAGE: 'image',
    LINK: 'link',
    RICH: 'rich',
    TWEET: 'tweet',
    VIDEO: 'video',
});
exports.MessageFlags = Object.freeze({
    CROSSPOSTED: 1 << 0,
    IS_CROSSPOST: 1 << 1,
    SUPPRESS_EMBEDS: 1 << 2,
});
exports.MessageTypes = Object.freeze({
    BASE: -1,
    DEFAULT: 0,
    RECIPIENT_ADD: 1,
    RECIPIENT_REMOVE: 2,
    CALL: 3,
    CHANNEL_NAME_CHANGE: 4,
    CHANNEL_ICON_CHANGE: 5,
    CHANNEL_PINNED_MESSAGE: 6,
    GUILD_MEMBER_JOIN: 7,
    GUILD_PREMIUM_SUBSCRIPTION: 8,
    GUILD_PREMIUM_SUBSCRIPTION_TIER_1: 9,
    GUILD_PREMIUM_SUBSCRIPTION_TIER_2: 10,
    GUILD_PREMIUM_SUBSCRIPTION_TIER_3: 11,
    CHANNEL_FOLLOW_ADD: 12,
    LFG_DELIST_EXPIRED: 13,
    LFG_DELIST_COMPLETED: 14,
    LFG_DELIST_INACTIVE: 15,
    LFG_DELIST_UNKNOWN: 16,
    LFG_RELIST: 17,
});
exports.MessageTypesDeletable = Object.freeze({
    [exports.MessageTypes.BASE]: true,
    [exports.MessageTypes.DEFAULT]: true,
    [exports.MessageTypes.CHANNEL_PINNED_MESSAGE]: true,
    [exports.MessageTypes.GUILD_MEMBER_JOIN]: true,
    [exports.MessageTypes.GUILD_PREMIUM_SUBSCRIPTION]: true,
    [exports.MessageTypes.GUILD_PREMIUM_SUBSCRIPTION_TIER_1]: true,
    [exports.MessageTypes.GUILD_PREMIUM_SUBSCRIPTION_TIER_2]: true,
    [exports.MessageTypes.GUILD_PREMIUM_SUBSCRIPTION_TIER_3]: true,
    [exports.MessageTypes.CHANNEL_FOLLOW_ADD]: true,
});
exports.MfaLevels = Object.freeze({
    NONE: 0,
    ELEVATED: 1,
});
exports.OverwriteTypes = Object.freeze({
    MEMBER: 'member',
    ROLE: 'role',
});
exports.Permissions = Object.freeze({
    NONE: 0,
    CREATE_INSTANT_INVITE: 1 << 0,
    KICK_MEMBERS: 1 << 1,
    BAN_MEMBERS: 1 << 2,
    ADMINISTRATOR: 1 << 3,
    MANAGE_CHANNELS: 1 << 4,
    MANAGE_GUILD: 1 << 5,
    ADD_REACTIONS: 1 << 6,
    VIEW_AUDIT_LOG: 1 << 7,
    PRIORITY_SPEAKER: 1 << 8,
    STREAM: 1 << 9,
    VIEW_CHANNEL: 1 << 10,
    SEND_MESSAGES: 1 << 11,
    SEND_TTS_MESSAGES: 1 << 12,
    MANAGE_MESSAGES: 1 << 13,
    EMBED_LINKS: 1 << 14,
    ATTACH_FILES: 1 << 15,
    READ_MESSAGE_HISTORY: 1 << 16,
    MENTION_EVERYONE: 1 << 17,
    USE_EXTERNAL_EMOJIS: 1 << 18,
    // unreleased feature 1 << 19
    CONNECT: 1 << 20,
    SPEAK: 1 << 21,
    MUTE_MEMBERS: 1 << 22,
    DEAFEN_MEMBERS: 1 << 23,
    MOVE_MEMBERS: 1 << 24,
    USE_VAD: 1 << 25,
    CHANGE_NICKNAME: 1 << 26,
    CHANGE_NICKNAMES: 1 << 27,
    MANAGE_ROLES: 1 << 28,
    MANAGE_WEBHOOKS: 1 << 29,
    MANAGE_EMOJIS: 1 << 30,
});
exports.PERMISSIONS_ALL = Object.values(exports.Permissions).reduce((permissions, permission) => permissions | permission, exports.Permissions.NONE);
exports.PERMISSIONS_ALL_TEXT = [
    exports.Permissions.ADD_REACTIONS,
    exports.Permissions.SEND_MESSAGES,
    exports.Permissions.SEND_TTS_MESSAGES,
    exports.Permissions.MANAGE_MESSAGES,
    exports.Permissions.EMBED_LINKS,
    exports.Permissions.ATTACH_FILES,
    exports.Permissions.READ_MESSAGE_HISTORY,
    exports.Permissions.MENTION_EVERYONE,
    exports.Permissions.USE_EXTERNAL_EMOJIS,
].reduce((permissions, permission) => permissions | permission, exports.Permissions.NONE);
exports.PERMISSIONS_ALL_VOICE = [
    exports.Permissions.PRIORITY_SPEAKER,
    exports.Permissions.STREAM,
    exports.Permissions.CONNECT,
    exports.Permissions.SPEAK,
    exports.Permissions.MUTE_MEMBERS,
    exports.Permissions.DEAFEN_MEMBERS,
    exports.Permissions.MOVE_MEMBERS,
    exports.Permissions.USE_VAD,
].reduce((permissions, permission) => permissions | permission, exports.Permissions.NONE);
exports.PERMISSIONS_DEFAULT = [
    exports.Permissions.CREATE_INSTANT_INVITE,
    exports.Permissions.CHANGE_NICKNAME,
    exports.Permissions.VIEW_CHANNEL,
    exports.Permissions.ADD_REACTIONS,
    exports.Permissions.SEND_MESSAGES,
    exports.Permissions.SEND_TTS_MESSAGES,
    exports.Permissions.EMBED_LINKS,
    exports.Permissions.ATTACH_FILES,
    exports.Permissions.READ_MESSAGE_HISTORY,
    exports.Permissions.MENTION_EVERYONE,
    exports.Permissions.USE_EXTERNAL_EMOJIS,
    exports.Permissions.STREAM,
    exports.Permissions.CONNECT,
    exports.Permissions.SPEAK,
    exports.Permissions.USE_VAD,
].reduce((permissions, permission) => permissions | permission, exports.Permissions.NONE);
exports.PERMISSIONS_LURKER = [
    exports.Permissions.VIEW_CHANNEL,
    exports.Permissions.READ_MESSAGE_HISTORY,
].reduce((permissions, permission) => permissions | permission, exports.Permissions.NONE);
exports.PlatformTypes = Object.freeze({
    BATTLENET: 'battlenet',
    FACEBOOK: 'facebook',
    LEAGUE_OF_LEGENDS: 'leagueoflegends',
    REDDIT: 'reddit',
    SKYPE: 'skype',
    SPOTIFY: 'spotify',
    STEAM: 'steam',
    TWITCH: 'twitch',
    TWITTER: 'twitter',
    YUOTUBE: 'youtube',
    XBOX: 'xbox',
});
exports.PremiumGuildTiers = Object.freeze({
    NONE: 0,
    TIER_1: 1,
    TIER_2: 2,
    TIER_3: 3,
});
exports.PremiumGuildTierNames = Object.freeze({
    [exports.PremiumGuildTiers.NONE]: '',
    [exports.PremiumGuildTiers.TIER_1]: 'Level 1',
    [exports.PremiumGuildTiers.TIER_2]: 'Level 2',
    [exports.PremiumGuildTiers.TIER_3]: 'Level 3',
});
exports.PremiumGuildSubscriptionsRequired = Object.freeze({
    [exports.PremiumGuildTiers.NONE]: 0,
    [exports.PremiumGuildTiers.TIER_1]: 2,
    [exports.PremiumGuildTiers.TIER_2]: 10,
    [exports.PremiumGuildTiers.TIER_3]: 20,
});
exports.PremiumGuildLimits = Object.freeze({
    [exports.PremiumGuildTiers.NONE]: Object.freeze({
        attachment: exports.MAX_ATTACHMENT_SIZE,
        bitrate: exports.MAX_BITRATE,
        emoji: exports.MAX_EMOJI_SLOTS,
    }),
    [exports.PremiumGuildTiers.TIER_1]: Object.freeze({
        attachment: exports.MAX_ATTACHMENT_SIZE,
        bitrate: 128000,
        emoji: 100,
    }),
    [exports.PremiumGuildTiers.TIER_2]: Object.freeze({
        attachment: exports.MAX_ATTACHMENT_SIZE_PREMIUM,
        bitrate: 256000,
        emoji: 150,
    }),
    [exports.PremiumGuildTiers.TIER_3]: Object.freeze({
        attachment: exports.MAX_ATTACHMENT_SIZE_PREMIUM * 2,
        bitrate: 384000,
        emoji: 250,
    }),
});
exports.PremiumTypes = Object.freeze({
    NONE: 0,
    TIER_1: 1,
    TIER_2: 2,
});
exports.RelationshipTypes = Object.freeze({
    NONE: 0,
    FRIEND: 1,
    BLOCKED: 2,
    PENDING_INCOMING: 3,
    PENDING_OUTGOING: 4,
    IMPLICIT: 5,
});
exports.SpecialUrls = detritus_utils_1.Tools.URIEncodeWrap({
    SPOTIFY_TRACK: (trackId) => `https://open.spotify.com/track/${trackId}`,
    YOUTUBE_VIDEO: (videoId) => `https://youtu.be/${videoId}`,
});
exports.SystemChannelFlags = Object.freeze({
    SUPPRESS_JOIN_NOTIFICATIONS: 1 << 0,
    SUPPRESS_PREMIUM_SUBSCRIPTIONS: 1 << 0,
});
exports.SystemMessages = Object.freeze({
    CallMissed: 'You missed a call from :user:.',
    CallStarted: ':user: started a call.',
    ChannelFollowAdd: ':user: has added **:webhookName:** to this channel',
    ChannelIconChange: ':user: changed the channel name: **:name:**',
    ChannelNameChange: ':user: changed the channel icon.',
    PinnedMessage: ':user: pinned a message to this channel.',
    RecipientAdd: ':user: added :user2: to the group.',
    RecipientRemove: ':user: removed :user2: from the group.',
    RecipientRemoveSelf: ':user: left the group.',
    GuildMemberJoin: [
        ":user: just joined the server - glhf!",
        ":user: just joined. Everyone, look busy!",
        ":user: just joined. Can I get a heal?",
        ":user: joined your party.",
        ":user: joined. You must construct additional pylons.",
        "Ermagherd. :user: is here.",
        "Welcome, :user:. Stay awhile and listen.",
        "Welcome, :user:. We were expecting you ( ͡° ͜ʖ ͡°)",
        "Welcome, :user:. We hope you brought pizza.",
        "Welcome :user:. Leave your weapons by the door.",
        "A wild :user: appeared.",
        "Swoooosh. :user: just landed.",
        "Brace yourselves. :user: just joined the server.",
        ":user: just joined. Hide your bananas.",
        ":user: just arrived. Seems OP - please nerf.",
        ":user: just slid into the server.",
        "A :user: has spawned in the server.",
        "Big :user: showed up!",
        "Where’s :user:? In the server!",
        ":user: hopped into the server. Kangaroo!!",
        ":user: just showed up. Hold my beer.",
        "Challenger approaching - :user: has appeared!",
        "It's a bird! It's a plane! Nevermind, it's just :user:.",
        "It's :user:! Praise the sun! \\\\[T]/",
        "Never gonna give :user: up. Never gonna let :user: down.",
        "Ha! :user: has joined! You activated my trap card!",
        "Cheers, love! :user:'s here!",
        "Hey! Listen! :user: has joined!",
        "We've been expecting you :user:",
        "It's dangerous to go alone, take :user:!",
        ":user: has joined the server! It's super effective!",
        "Cheers, love! :user: is here!",
        ":user: is here, as the prophecy foretold.",
        ":user: has arrived. Party's over.",
        "Ready player :user:",
        ":user: is here to kick butt and chew bubblegum. And :user: is all out of gum.",
        "Hello. Is it :user: you're looking for?",
        ":user: has joined. Stay a while and listen!",
        "Roses are red, violets are blue, :user: joined this server with you",
    ],
    GuildMemberSubscribed: ':user: just boosted the server!',
    GuildMemberSubscribedAchievedTier: ':user: just boosted the server! :guild: has achieved **:premiumTier:!**',
});
exports.TeamMembershipStates = Object.freeze({
    BASE: 0,
    INVITED: 1,
    ACCEPTED: 2,
});
exports.TeamPayoutAccountStatuses = Object.freeze({
    BASE: -2,
    UNSUBMITTED: -1,
    PENDING: 2,
    ACTION_REQUIRED: 3,
    ACTIVE: 4,
    BLOCKED: 5,
    SUSPENDED: 6,
});
exports.TYPING_TIMEOUT = 10000;
exports.UserFlags = Object.freeze({
    STAFF: 1 << 0,
    PARTNER: 1 << 1,
    HYPESQUAD: 1 << 2,
    BUG_HUNTER: 1 << 3,
    MFA_SMS: 1 << 4,
    PREMIUM_PROMO_DISMISSED: 1 << 5,
    HYPESQUAD_ONLINE_HOUSE_1: 1 << 6,
    HYPESQUAD_ONLINE_HOUSE_2: 1 << 7,
    HYPESQUAD_ONLINE_HOUSE_3: 1 << 8,
    PREMIUM_EARLY_SUPPORTER: 1 << 9,
    TEAM_USER: 1 << 10,
});
exports.UserRequiredActions = detritus_utils_1.Tools.normalize({
    AGREEMENTS: null,
    REQUIRE_CAPTCHA: null,
    REQUIRE_VERIFIED_EMAIL: null,
    REQUIRE_VERIFIED_PHONE: null,
});
exports.VerificationLevels = Object.freeze({
    NONE: 0,
    LOW: 1,
    MEDIUM: 2,
    HIGH: 3,
    VERY_HIGH: 4,
});
exports.GuildWidgetStyles = Object.freeze({
    BANNER_1: 'banner1',
    BANNER_2: 'banner2',
    BANNER_3: 'banner3',
    BANNER_4: 'banner4',
    SHIELD: 'shield',
});
exports.DiscordKeys = Object.freeze({
    ACCESS_TYPE: 'access_type',
    ACCOUNT: 'account',
    ACTION_TYPE: 'action_type',
    ACTIVE: 'active',
    ACTIVITIES: 'activities',
    ACTIVITY: 'activity',
    AFK_CHANNEL_ID: 'afk_channel_id',
    AFK_TIMEOUT: 'afk_timeout',
    ALIASES: 'aliases',
    ALLOW: 'allow',
    ANIMATED: 'animated',
    APPLICATION: 'application',
    APPLICATION_ID: 'application_id',
    APPROXIMATE_MEMBER_COUNT: 'approximate_member_count',
    APPROXIMATE_PRESENCE_COUNT: 'approximate_presence_count',
    ASSETS: 'assets',
    ATTACHMENTS: 'attachments',
    AUTHOR: 'author',
    AVAILABLE: 'available',
    AVATAR: 'avatar',
    BANNER: 'banner',
    BITRATE: 'bitrate',
    BOT: 'bot',
    BOT_PUBLIC: 'bot_public',
    BOT_REQUIRE_CODE_GRANT: 'bot_require_code_grant',
    BOX_ART: 'box_art',
    CALL: 'call',
    CAROUSEL_ITEMS: 'carousel_items',
    CATEGORY: 'category',
    CHANGES: 'changes',
    CHANNEL: 'channel',
    CHANNELS: 'channels',
    CHANNEL_ID: 'channel_id',
    CLIENT: 'client',
    CLIENT_INFO: 'client_info',
    CLIENT_STATUS: 'client_status',
    CODE: 'code',
    COLOR: 'color',
    CONNECTED_ACCOUNTS: 'connected_accounts',
    CONTENT: 'content',
    CONTENT_RATING: 'content_rating',
    CONTENT_RATING_AGENCY: 'content_rating_agency',
    COUNT: 'count',
    COVER_IMAGE: 'cover_image',
    CREATED_AT: 'created_at',
    CURRENCY: 'currency',
    CUSTOM: 'custom',
    DEAF: 'deaf',
    DEFAULT_MESSAGE_NOTIFICATIONS: 'default_message_notifications',
    DELETE_MEMBER_DAYS: 'delete_member_days',
    DENY: 'deny',
    DEPENDENT_SKU_ID: 'dependent_sku_id',
    DEPRECATED: 'deprecated',
    DESCRIPTION: 'description',
    DESKTOP: 'desktop',
    DETAILS: 'details',
    DEVELOPERS: 'developers',
    DISCRIMINATOR: 'discriminator',
    DISTRIBUTOR: 'distributor',
    EDITED_TIMESTAMP: 'edited_timestamp',
    EMAIL: 'email',
    EMBEDS: 'embeds',
    EMBED_CHANNEL_ID: 'embed_channel_id',
    EMBED_ENABLED: 'embed_enabled',
    EMOJI: 'emoji',
    EMOJIS: 'emojis',
    ENABLED: 'enabled',
    END: 'end',
    ENDED: 'ended',
    ENDED_TIMESTAMP: 'ended_timestamp',
    ENTITLEMENT_BRANCH_ID: 'entitlement_branch_id',
    EULA_ID: 'eula_id',
    EXECUTABLES: 'executables',
    EXPIRES_AT: 'expires_at',
    EXPIRE_BEHAVIOR: 'expire_behavior',
    EXPIRE_GRACE_PERIOD: 'expire_grace_period',
    EXPLICIT_CONTENT_FILTER: 'explicit_content_filter',
    FEATURES: 'features',
    FIELDS: 'fields',
    FLAGS: 'flags',
    FILENAME: 'filename',
    FOOTER: 'footer',
    FRIEND_SYNC: 'friend_sync',
    GAME: 'game',
    GAME_ID: 'game_id',
    GENRES: 'genres',
    GUILD: 'guild',
    GUILD_ID: 'guild_id',
    GUILD_IDS: 'guild_ids',
    HEADER_BACKGROUND: 'header_background',
    HEADER_LOGO_DARK_THEME: 'header_logo_dark_theme',
    HEADER_LOGO_LIGHT_THEME: 'header_logo_light_theme',
    HEIGHT: 'height',
    HERO_BACKGROUND: 'hero_background',
    HERO_VIDEO: 'hero_video',
    HOIST: 'hoist',
    HOISTED_ROLE: 'hoisted_role',
    ICON: 'icon',
    ICON_URL: 'icon_url',
    ID: 'id',
    IMAGE: 'image',
    INLINE: 'inline',
    INSTANCE: 'instance',
    INTEGRATIONS: 'integrations',
    INTERVAL: 'interval',
    INTERVAL_COUNT: 'interval_count',
    INVITER: 'inviter',
    IS_PARTIAL: 'is_partial',
    JOIN: 'join',
    JOINED_AT: 'joined_at',
    KEY: 'key',
    LARGE: 'large',
    LARGE_IMAGE: 'large_image',
    LARGE_TEXT: 'large_text',
    LAST_MESSAGE_ID: 'last_message_id',
    LAST_MODIFIED: 'last_modified',
    LAST_PIN_TIMESTAMP: 'last_pin_timestamp',
    LAZY: 'lazy',
    LEGAL_NOTICE: 'legal_notice',
    LOCALE: 'locale',
    LOCALES: 'locales',
    MANAGED: 'managed',
    MANIFEST_LABELS: 'manifest_labels',
    MATCH: 'match',
    MAX_AGE: 'max_age',
    MAX_MEMBERS: 'max_members',
    MAX_PRESENCES: 'max_presences',
    MAX_USES: 'max_uses',
    ME: 'me',
    MEMBER: 'member',
    MEMBERS: 'members',
    MEMBERSHIP_STATE: 'membership_state',
    MEMBERS_REMOVED: 'members_removed',
    MEMBER_COUNT: 'member_count',
    MENTIONABLE: 'mentionable',
    MENTIONS: 'mentions',
    MENTION_CHANNELS: 'mention_channels',
    MENTION_EVERYONE: 'mention_everyone',
    MENTION_ROLES: 'mention_roles',
    MESSAGE_ID: 'message_id',
    MESSAGE_REFERENCE: 'message_reference',
    METADATA: 'metadata',
    MFA_ENABLED: 'mfa_enabled',
    MFA_LEVEL: 'mfa_level',
    MIME_TYPE: 'mime_type',
    MOBILE: 'mobile',
    MUTE: 'mute',
    MUTUAL_GUILDS: 'mutual_guilds',
    NAME: 'name',
    NEW_VALUE: 'new_value',
    NICK: 'nick',
    NICKS: 'nicks',
    NONCE: 'nonce',
    NSFW: 'nsfw',
    OLD_VALUE: 'old_value',
    OPTIMAL: 'optimal',
    OPTIONS: 'options',
    OS: 'os',
    OVERLAY: 'overlay',
    OVERLAY_COMPATIBILITY_HOOK: 'overlay_compatibility_hook',
    OWNER: 'owner',
    OWNER_ID: 'owner_id',
    OWNER_USER_ID: 'owner_user_id',
    PARENT_ID: 'parent_id',
    PARTICIPANTS: 'participants',
    PARTY: 'party',
    PARTY_ID: 'party_id',
    PAYOUT_ACCOUNT_STATUS: 'payout_account_status',
    PERMISSIONS: 'permissions',
    PERMISSION_OVERWRITES: 'permission_overwrites',
    PHONE: 'phone',
    PINNED: 'pinned',
    PLATFORM: 'platform',
    POSITION: 'position',
    PREFERRED_LOCALE: 'preferred_locale',
    PREMIUM: 'premium',
    PREMIUM_GUILD_SINCE: 'premium_guild_since',
    PREMIUM_SINCE: 'premium_since',
    PREMIUM_SUBSCRIPTION_COUNT: 'premium_subscription_count',
    PREMIUM_TIER: 'premium_tier',
    PREMIUM_TYPE: 'premium_type',
    PRESENCES: 'presences',
    PREVIEW_VIDEO: 'preview_video',
    PRICE: 'price',
    PRIMARY_SKU_ID: 'primary_sku_id',
    PROVIDER: 'provider',
    PROXY_ICON_URL: 'proxy_icon_url',
    PROXY_URL: 'proxy_url',
    PUBLISHERS: 'publishers',
    RATE_LIMIT_PER_USER: 'rate_limit_per_user',
    REACTIONS: 'reactions',
    REASON: 'reason',
    RECIPIENTS: 'recipients',
    REDEEMED: 'redeemed',
    REDIRECT_URIS: 'redirect_uris',
    REFERENCE_ID: 'reference_id',
    REGION: 'region',
    RELEASE_DATE: 'release_date',
    REQUIRE_COLONS: 'require_colons',
    REVOKED: 'revoked',
    RINGING: 'ringing',
    ROLES: 'roles',
    ROLE_ID: 'role_id',
    RPC_APPLICATION_STATE: 'rpc_application_state',
    RPC_ORIGINS: 'rpc_origins',
    SECRET: 'secret',
    SECRETS: 'secrets',
    SELF_DEAF: 'self_deaf',
    SELF_MUTE: 'self_mute',
    SELF_STREAM: 'self_stream',
    SELF_VIDEO: 'self_video',
    SESSION_ID: 'session_id',
    SHOW_ACTIVITY: 'show_activity',
    SHOW_AGE_GATE: 'show_age_gate',
    SIZE: 'size',
    SKU: 'sku',
    SKU_ID: 'sku_id',
    SLUG: 'slug',
    SMALL_IMAGE: 'small_image',
    SMALL_TEXT: 'small_text',
    SPECTATE: 'spectate',
    SPLASH: 'splash',
    START: 'start',
    STARTED: 'started',
    STATE: 'state',
    STATUS: 'status',
    STOPPED: 'stopped',
    STORE_APPLICATION_STATE: 'store_application_state',
    STORE_LISTING: 'store_listing',
    SUBSCRIPTION_PLAN: 'subscription_plan',
    SUBSCRIPTION_PLAN_ID: 'subscription_plan_id',
    SUBTARGET: 'subtarget',
    SUMMARY: 'summary',
    SUPPRESS: 'suppress',
    SYNCED_AT: 'synced_at',
    SYNCING: 'syncing',
    SYNC_ID: 'sync_id',
    SYSTEM_CHANNEL_FLAGS: 'system_channel_flags',
    SYSTEM_CHANNEL_ID: 'system_channel_id',
    SYSTEM_REQUIREMENTS: 'system_requirements',
    TAGLINE: 'tagline',
    TARGET: 'target',
    TARGET_ID: 'target_id',
    TARGET_USER: 'target_user',
    TARGET_USER_TYPE: 'target_user_type',
    TAX_INCLUSIVE: 'tax_inclusive',
    TEAM: 'team',
    TEAM_ID: 'team_id',
    TEMPORARY: 'temporary',
    TEXT: 'text',
    THIRD_PARTY_SKUS: 'third_party_skus',
    THUMBNAIL: 'thumbnail',
    TIMESTAMP: 'timestamp',
    TIMESTAMPS: 'timestamps',
    TITLE: 'title',
    TOKEN: 'token',
    TOPIC: 'topic',
    TTS: 'tts',
    TYPE: 'type',
    UNAVAILABLE: 'unavailable',
    URL: 'url',
    USER: 'user',
    USERNAME: 'username',
    USER_ID: 'user_id',
    USER_LIMIT: 'user_limit',
    USES: 'uses',
    VALUE: 'value',
    VANITY_URL_CODE: 'vanity_url_code',
    VERIFICATION_LEVEL: 'verification_level',
    VERIFIED: 'verified',
    VERIFY_KEY: 'verify_key',
    VERSION: 'version',
    VIDEO: 'video',
    VIP: 'vip',
    VISIBILITY: 'visibility',
    VOICE_STATES: 'voice_states',
    WEB: 'web',
    WEBHOOK_ID: 'webhook_id',
    WIDGET_CHANNEL_ID: 'widget_channel_id',
    WIDGET_ENABLED: 'widget_enabled',
    WIDTH: 'width',
    YOUTUBE_TRAILER_VIDEO_ID: 'youtube_trailer_video_id',
});
exports.DetritusKeys = Object.freeze({
    [exports.DiscordKeys.ACCESS_TYPE]: 'accessType',
    [exports.DiscordKeys.ACCOUNT]: 'account',
    [exports.DiscordKeys.ACTION_TYPE]: 'actionType',
    [exports.DiscordKeys.ACTIVE]: 'active',
    [exports.DiscordKeys.ACTIVITIES]: 'activities',
    [exports.DiscordKeys.ACTIVITY]: 'activity',
    [exports.DiscordKeys.AFK_CHANNEL_ID]: 'afkChannelId',
    [exports.DiscordKeys.AFK_TIMEOUT]: 'afkTimeout',
    [exports.DiscordKeys.ALIASES]: 'aliases',
    [exports.DiscordKeys.ALLOW]: 'allow',
    [exports.DiscordKeys.ANIMATED]: 'animated',
    [exports.DiscordKeys.APPLICATION]: 'application',
    [exports.DiscordKeys.APPLICATION_ID]: 'applicationId',
    [exports.DiscordKeys.APPROXIMATE_MEMBER_COUNT]: 'approximateMemberCount',
    [exports.DiscordKeys.APPROXIMATE_PRESENCE_COUNT]: 'approximatePresenceCount',
    [exports.DiscordKeys.ASSETS]: 'assets',
    [exports.DiscordKeys.ATTACHMENTS]: 'attachments',
    [exports.DiscordKeys.AUTHOR]: 'author',
    [exports.DiscordKeys.AVAILABLE]: 'available',
    [exports.DiscordKeys.AVATAR]: 'avatar',
    [exports.DiscordKeys.BANNER]: 'banner',
    [exports.DiscordKeys.BITRATE]: 'bitrate',
    [exports.DiscordKeys.BOT]: 'bot',
    [exports.DiscordKeys.BOT_PUBLIC]: 'botPublic',
    [exports.DiscordKeys.BOT_REQUIRE_CODE_GRANT]: 'botRequireCodeGrant',
    [exports.DiscordKeys.BOX_ART]: 'boxArt',
    [exports.DiscordKeys.CALL]: 'call',
    [exports.DiscordKeys.CAROUSEL_ITEMS]: 'carouselItems',
    [exports.DiscordKeys.CATEGORY]: 'category',
    [exports.DiscordKeys.CHANGES]: 'changes',
    [exports.DiscordKeys.CHANNEL]: 'channel',
    [exports.DiscordKeys.CHANNELS]: 'channels',
    [exports.DiscordKeys.CHANNEL_ID]: 'channelId',
    [exports.DiscordKeys.CLIENT]: 'client',
    [exports.DiscordKeys.CLIENT_INFO]: 'clientInfo',
    [exports.DiscordKeys.CLIENT_STATUS]: 'clientStatus',
    [exports.DiscordKeys.CODE]: 'code',
    [exports.DiscordKeys.COLOR]: 'color',
    [exports.DiscordKeys.CONNECTED_ACCOUNTS]: 'connectedAccounts',
    [exports.DiscordKeys.CONTENT]: 'content',
    [exports.DiscordKeys.CONTENT_RATING]: 'contentRating',
    [exports.DiscordKeys.CONTENT_RATING_AGENCY]: 'contentRatingAgency',
    [exports.DiscordKeys.COUNT]: 'count',
    [exports.DiscordKeys.COVER_IMAGE]: 'coverImage',
    [exports.DiscordKeys.CREATED_AT]: 'createdAt',
    [exports.DiscordKeys.CURRENCY]: 'currency',
    [exports.DiscordKeys.CUSTOM]: 'custom',
    [exports.DiscordKeys.DEAF]: 'deaf',
    [exports.DiscordKeys.DEFAULT_MESSAGE_NOTIFICATIONS]: 'defaultMessageNotifications',
    [exports.DiscordKeys.DELETE_MEMBER_DAYS]: 'deleteMemberDays',
    [exports.DiscordKeys.DENY]: 'deny',
    [exports.DiscordKeys.DEPENDENT_SKU_ID]: 'dependentSkuId',
    [exports.DiscordKeys.DEPRECATED]: 'deprecated',
    [exports.DiscordKeys.DESCRIPTION]: 'description',
    [exports.DiscordKeys.DESKTOP]: 'desktop',
    [exports.DiscordKeys.DETAILS]: 'details',
    [exports.DiscordKeys.DEVELOPERS]: 'developers',
    [exports.DiscordKeys.DISCRIMINATOR]: 'discriminator',
    [exports.DiscordKeys.DISTRIBUTOR]: 'distributor',
    [exports.DiscordKeys.EDITED_TIMESTAMP]: 'editedTimestamp',
    [exports.DiscordKeys.EMAIL]: 'email',
    [exports.DiscordKeys.EMBEDS]: 'embeds',
    [exports.DiscordKeys.EMBED_CHANNEL_ID]: 'embedChannelId',
    [exports.DiscordKeys.EMBED_ENABLED]: 'embedEnabled',
    [exports.DiscordKeys.EMOJI]: 'emoji',
    [exports.DiscordKeys.EMOJIS]: 'emojis',
    [exports.DiscordKeys.ENABLED]: 'enabled',
    [exports.DiscordKeys.END]: 'end',
    [exports.DiscordKeys.ENDED]: 'ended',
    [exports.DiscordKeys.ENDED_TIMESTAMP]: 'endedTimestamp',
    [exports.DiscordKeys.ENTITLEMENT_BRANCH_ID]: 'entitlementBranchId',
    [exports.DiscordKeys.EULA_ID]: 'eulaId',
    [exports.DiscordKeys.EXECUTABLES]: 'executables',
    [exports.DiscordKeys.EXPIRES_AT]: 'expiresAt',
    [exports.DiscordKeys.EXPIRE_BEHAVIOR]: 'expireBehavior',
    [exports.DiscordKeys.EXPIRE_GRACE_PERIOD]: 'expireGracePeriod',
    [exports.DiscordKeys.EXPLICIT_CONTENT_FILTER]: 'explicitContentFilter',
    [exports.DiscordKeys.FEATURES]: 'features',
    [exports.DiscordKeys.FIELDS]: 'fields',
    [exports.DiscordKeys.FLAGS]: 'flags',
    [exports.DiscordKeys.FILENAME]: 'filename',
    [exports.DiscordKeys.FOOTER]: 'footer',
    [exports.DiscordKeys.FRIEND_SYNC]: 'friendSync',
    [exports.DiscordKeys.GAME]: 'game',
    [exports.DiscordKeys.GAME_ID]: 'gameId',
    [exports.DiscordKeys.GENRES]: 'genres',
    [exports.DiscordKeys.GUILD]: 'guild',
    [exports.DiscordKeys.GUILD_ID]: 'guildId',
    [exports.DiscordKeys.GUILD_IDS]: 'guildIds',
    [exports.DiscordKeys.HEADER_BACKGROUND]: 'headerBackground',
    [exports.DiscordKeys.HEADER_LOGO_DARK_THEME]: 'headerLogoDarkTheme',
    [exports.DiscordKeys.HEADER_LOGO_LIGHT_THEME]: 'headerLogoLightTheme',
    [exports.DiscordKeys.HEIGHT]: 'height',
    [exports.DiscordKeys.HERO_BACKGROUND]: 'heroBackground',
    [exports.DiscordKeys.HERO_VIDEO]: 'heroVideo',
    [exports.DiscordKeys.HOIST]: 'hoist',
    [exports.DiscordKeys.HOISTED_ROLE]: 'hoistedRole',
    [exports.DiscordKeys.ICON]: 'icon',
    [exports.DiscordKeys.ICON_URL]: 'iconUrl',
    [exports.DiscordKeys.ID]: 'id',
    [exports.DiscordKeys.IMAGE]: 'image',
    [exports.DiscordKeys.INLINE]: 'inline',
    [exports.DiscordKeys.INSTANCE]: 'instance',
    [exports.DiscordKeys.INTEGRATIONS]: 'integrations',
    [exports.DiscordKeys.INTERVAL]: 'interval',
    [exports.DiscordKeys.INTERVAL_COUNT]: 'intervalCount',
    [exports.DiscordKeys.INVITER]: 'inviter',
    [exports.DiscordKeys.IS_PARTIAL]: 'isPartial',
    [exports.DiscordKeys.JOIN]: 'join',
    [exports.DiscordKeys.JOINED_AT]: 'joinedAt',
    [exports.DiscordKeys.KEY]: 'key',
    [exports.DiscordKeys.LARGE]: 'large',
    [exports.DiscordKeys.LARGE_IMAGE]: 'largeImage',
    [exports.DiscordKeys.LARGE_TEXT]: 'largeText',
    [exports.DiscordKeys.LAST_MESSAGE_ID]: 'lastMessageId',
    [exports.DiscordKeys.LAST_MODIFIED]: 'lastModified',
    [exports.DiscordKeys.LAST_PIN_TIMESTAMP]: 'lastPinTimestamp',
    [exports.DiscordKeys.LAZY]: 'lazy',
    [exports.DiscordKeys.LEGAL_NOTICE]: 'legalNotice',
    [exports.DiscordKeys.LOCALE]: 'locale',
    [exports.DiscordKeys.LOCALES]: 'locales',
    [exports.DiscordKeys.MANAGED]: 'managed',
    [exports.DiscordKeys.MANIFEST_LABELS]: 'manifestLabels',
    [exports.DiscordKeys.MATCH]: 'match',
    [exports.DiscordKeys.MAX_AGE]: 'maxAge',
    [exports.DiscordKeys.MAX_MEMBERS]: 'maxMembers',
    [exports.DiscordKeys.MAX_PRESENCES]: 'maxPresences',
    [exports.DiscordKeys.MAX_USES]: 'maxUses',
    [exports.DiscordKeys.ME]: 'me',
    [exports.DiscordKeys.MEMBER]: 'member',
    [exports.DiscordKeys.MEMBERS]: 'members',
    [exports.DiscordKeys.MEMBERSHIP_STATE]: 'membershipState',
    [exports.DiscordKeys.MEMBERS_REMOVED]: 'membersRemoved',
    [exports.DiscordKeys.MEMBER_COUNT]: 'memberCount',
    [exports.DiscordKeys.MENTIONABLE]: 'mentionable',
    [exports.DiscordKeys.MENTIONS]: 'mentions',
    [exports.DiscordKeys.MENTION_CHANNELS]: 'mentionChannels',
    [exports.DiscordKeys.MENTION_EVERYONE]: 'mentionEveryone',
    [exports.DiscordKeys.MENTION_ROLES]: 'mentionRoles',
    [exports.DiscordKeys.MESSAGE_ID]: 'messageId',
    [exports.DiscordKeys.MESSAGE_REFERENCE]: 'messageReference',
    [exports.DiscordKeys.METADATA]: 'metadata',
    [exports.DiscordKeys.MFA_ENABLED]: 'mfaEnabled',
    [exports.DiscordKeys.MFA_LEVEL]: 'mfaLevel',
    [exports.DiscordKeys.MIME_TYPE]: 'mimeType',
    [exports.DiscordKeys.MOBILE]: 'mobile',
    [exports.DiscordKeys.MUTE]: 'mute',
    [exports.DiscordKeys.MUTUAL_GUILDS]: 'mutualGuilds',
    [exports.DiscordKeys.NAME]: 'name',
    [exports.DiscordKeys.NEW_VALUE]: 'newValue',
    [exports.DiscordKeys.NICK]: 'nick',
    [exports.DiscordKeys.NICKS]: 'nicks',
    [exports.DiscordKeys.NONCE]: 'nonce',
    [exports.DiscordKeys.NSFW]: 'nsfw',
    [exports.DiscordKeys.OLD_VALUE]: 'oldValue',
    [exports.DiscordKeys.OPTIMAL]: 'optimal',
    [exports.DiscordKeys.OPTIONS]: 'options',
    [exports.DiscordKeys.OS]: 'os',
    [exports.DiscordKeys.OVERLAY]: 'overlay',
    [exports.DiscordKeys.OVERLAY_COMPATIBILITY_HOOK]: 'overlayCompatibilityHook',
    [exports.DiscordKeys.OWNER]: 'owner',
    [exports.DiscordKeys.OWNER_ID]: 'ownerId',
    [exports.DiscordKeys.OWNER_USER_ID]: 'ownerUserId',
    [exports.DiscordKeys.PARENT_ID]: 'parentId',
    [exports.DiscordKeys.PARTICIPANTS]: 'participants',
    [exports.DiscordKeys.PARTY]: 'party',
    [exports.DiscordKeys.PARTY_ID]: 'partyId',
    [exports.DiscordKeys.PAYOUT_ACCOUNT_STATUS]: 'payoutAccountStatus',
    [exports.DiscordKeys.PERMISSIONS]: 'permissions',
    [exports.DiscordKeys.PERMISSION_OVERWRITES]: 'permissionOverwrites',
    [exports.DiscordKeys.PHONE]: 'phone',
    [exports.DiscordKeys.PINNED]: 'pinned',
    [exports.DiscordKeys.PLATFORM]: 'platform',
    [exports.DiscordKeys.POSITION]: 'position',
    [exports.DiscordKeys.PREFERRED_LOCALE]: 'preferredLocale',
    [exports.DiscordKeys.PREMIUM]: 'premium',
    [exports.DiscordKeys.PREMIUM_GUILD_SINCE]: 'premiumGuildSince',
    [exports.DiscordKeys.PREMIUM_SINCE]: 'premiumSince',
    [exports.DiscordKeys.PREMIUM_SUBSCRIPTION_COUNT]: 'premiumSubscriptionCount',
    [exports.DiscordKeys.PREMIUM_TIER]: 'premiumTier',
    [exports.DiscordKeys.PREMIUM_TYPE]: 'premiumType',
    [exports.DiscordKeys.PRESENCES]: 'presences',
    [exports.DiscordKeys.PREVIEW_VIDEO]: 'previewVideo',
    [exports.DiscordKeys.PRICE]: 'price',
    [exports.DiscordKeys.PRIMARY_SKU_ID]: 'primarySkuId',
    [exports.DiscordKeys.PROVIDER]: 'provider',
    [exports.DiscordKeys.PROXY_ICON_URL]: 'proxyIconUrl',
    [exports.DiscordKeys.PROXY_URL]: 'proxyUrl',
    [exports.DiscordKeys.PUBLISHERS]: 'publishers',
    [exports.DiscordKeys.RATE_LIMIT_PER_USER]: 'rateLimitPerUser',
    [exports.DiscordKeys.REACTIONS]: 'reactions',
    [exports.DiscordKeys.REASON]: 'reason',
    [exports.DiscordKeys.RECIPIENTS]: 'recipients',
    [exports.DiscordKeys.REDEEMED]: 'redeemed',
    [exports.DiscordKeys.REDIRECT_URIS]: 'redirectUris',
    [exports.DiscordKeys.REFERENCE_ID]: 'referenceId',
    [exports.DiscordKeys.REGION]: 'region',
    [exports.DiscordKeys.RELEASE_DATE]: 'releaseDate',
    [exports.DiscordKeys.REQUIRE_COLONS]: 'requireColons',
    [exports.DiscordKeys.REVOKED]: 'revoked',
    [exports.DiscordKeys.RINGING]: 'ringing',
    [exports.DiscordKeys.ROLES]: 'roles',
    [exports.DiscordKeys.ROLE_ID]: 'roleId',
    [exports.DiscordKeys.RPC_APPLICATION_STATE]: 'rpcApplicationState',
    [exports.DiscordKeys.RPC_ORIGINS]: 'rpcOrigins',
    [exports.DiscordKeys.SECRET]: 'secret',
    [exports.DiscordKeys.SECRETS]: 'secrets',
    [exports.DiscordKeys.SELF_DEAF]: 'selfDeaf',
    [exports.DiscordKeys.SELF_MUTE]: 'selfMute',
    [exports.DiscordKeys.SELF_STREAM]: 'selfStream',
    [exports.DiscordKeys.SELF_VIDEO]: 'selfVideo',
    [exports.DiscordKeys.SESSION_ID]: 'sessionId',
    [exports.DiscordKeys.SHOW_ACTIVITY]: 'showActivity',
    [exports.DiscordKeys.SHOW_AGE_GATE]: 'showAgeGate',
    [exports.DiscordKeys.SIZE]: 'size',
    [exports.DiscordKeys.SKU]: 'sku',
    [exports.DiscordKeys.SKU_ID]: 'skuId',
    [exports.DiscordKeys.SLUG]: 'slug',
    [exports.DiscordKeys.SMALL_IMAGE]: 'smallImage',
    [exports.DiscordKeys.SMALL_TEXT]: 'smallText',
    [exports.DiscordKeys.SPECTATE]: 'spectate',
    [exports.DiscordKeys.SPLASH]: 'splash',
    [exports.DiscordKeys.START]: 'start',
    [exports.DiscordKeys.STARTED]: 'started',
    [exports.DiscordKeys.STATE]: 'state',
    [exports.DiscordKeys.STATUS]: 'status',
    [exports.DiscordKeys.STOPPED]: 'stopped',
    [exports.DiscordKeys.STORE_APPLICATION_STATE]: 'storeApplicationState',
    [exports.DiscordKeys.STORE_LISTING]: 'storeListing',
    [exports.DiscordKeys.SUBSCRIPTION_PLAN]: 'subscriptionPlan',
    [exports.DiscordKeys.SUBSCRIPTION_PLAN_ID]: 'subscriptionPlanId',
    [exports.DiscordKeys.SUBTARGET]: 'subtarget',
    [exports.DiscordKeys.SUMMARY]: 'summary',
    [exports.DiscordKeys.SUPPRESS]: 'suppress',
    [exports.DiscordKeys.SYNCED_AT]: 'syncedAt',
    [exports.DiscordKeys.SYNCING]: 'syncing',
    [exports.DiscordKeys.SYNC_ID]: 'syncId',
    [exports.DiscordKeys.SYSTEM_CHANNEL_FLAGS]: 'systemChannelFlags',
    [exports.DiscordKeys.SYSTEM_CHANNEL_ID]: 'systemChannelId',
    [exports.DiscordKeys.SYSTEM_REQUIREMENTS]: 'systemRequirements',
    [exports.DiscordKeys.TAGLINE]: 'tagline',
    [exports.DiscordKeys.TARGET]: 'target',
    [exports.DiscordKeys.TARGET_ID]: 'targetId',
    [exports.DiscordKeys.TARGET_USER]: 'targetUser',
    [exports.DiscordKeys.TARGET_USER_TYPE]: 'targetUserType',
    [exports.DiscordKeys.TAX_INCLUSIVE]: 'taxInclusive',
    [exports.DiscordKeys.TEAM]: 'team',
    [exports.DiscordKeys.TEAM_ID]: 'teamId',
    [exports.DiscordKeys.TEMPORARY]: 'temporary',
    [exports.DiscordKeys.TEXT]: 'text',
    [exports.DiscordKeys.THIRD_PARTY_SKUS]: 'thirdPartySkus',
    [exports.DiscordKeys.THUMBNAIL]: 'thumbnail',
    [exports.DiscordKeys.TIMESTAMP]: 'timestamp',
    [exports.DiscordKeys.TIMESTAMPS]: 'timestamps',
    [exports.DiscordKeys.TITLE]: 'title',
    [exports.DiscordKeys.TOKEN]: 'token',
    [exports.DiscordKeys.TOPIC]: 'topic',
    [exports.DiscordKeys.TTS]: 'tts',
    [exports.DiscordKeys.TYPE]: 'type',
    [exports.DiscordKeys.UNAVAILABLE]: 'unavailable',
    [exports.DiscordKeys.URL]: 'url',
    [exports.DiscordKeys.USER]: 'user',
    [exports.DiscordKeys.USERNAME]: 'username',
    [exports.DiscordKeys.USER_ID]: 'userId',
    [exports.DiscordKeys.USER_LIMIT]: 'userLimit',
    [exports.DiscordKeys.USES]: 'uses',
    [exports.DiscordKeys.VALUE]: 'value',
    [exports.DiscordKeys.VANITY_URL_CODE]: 'vanityUrlCode',
    [exports.DiscordKeys.VERIFICATION_LEVEL]: 'verificationLevel',
    [exports.DiscordKeys.VERIFIED]: 'verified',
    [exports.DiscordKeys.VERIFY_KEY]: 'verifyKey',
    [exports.DiscordKeys.VERSION]: 'version',
    [exports.DiscordKeys.VIDEO]: 'video',
    [exports.DiscordKeys.VIP]: 'vip',
    [exports.DiscordKeys.VISIBILITY]: 'visibility',
    [exports.DiscordKeys.VOICE_STATES]: 'voiceStates',
    [exports.DiscordKeys.WEB]: 'web',
    [exports.DiscordKeys.WEBHOOK_ID]: 'webhookId',
    [exports.DiscordKeys.WIDGET_CHANNEL_ID]: 'widgetChannelId',
    [exports.DiscordKeys.WIDGET_ENABLED]: 'widgetEnabled',
    [exports.DiscordKeys.WIDTH]: 'width',
    [exports.DiscordKeys.YOUTUBE_TRAILER_VIDEO_ID]: 'youtubeTrailerVideoId',
});
exports.COMMAND_RATELIMIT_TYPES = Object.freeze(Object.values(exports.CommandRatelimitTypes));
exports.DEFAULT_GROUP_DM_AVATARS = Object.freeze([
    '861ab526aa1fabb04c6b7da8074e3e21',
    'b8912961ea6ab32f0655d583bbc26b4f',
    '773616c3c8a7e21f8a774eb0d5625436',
    'f810dc5fedb7175c43a3389aa890534f',
    'e1fb24a120bdd003a84e021b16ec3bef',
    'b3150d5cef84b9e82128a1131684f287',
    '485a854d5171c8dc98088041626e6fea',
    '1531b79c2f2927945582023e1edaaa11',
]);
exports.IMAGE_FORMATS = Object.freeze(Object.values(exports.ImageFormats));
