// detritus-client/lib/utils/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
const url_1 = require("url");
const detritus_utils_1 = require("detritus-utils");
exports.Snowflake = detritus_utils_1.Snowflake;
const { guildIdToShardId } = detritus_utils_1.Tools;
exports.guildIdToShardId = guildIdToShardId;
const constants_1 = require("../constants");
const Markup = require("./markup");
exports.Markup = Markup;
const PermissionTools = require("./permissions");
exports.PermissionTools = PermissionTools;
__export(require("./embed"));
function addQuery(url, query) {
    if (query) {
        const params = new url_1.URLSearchParams();
        for (let key in query) {
            if (query[key] !== undefined) {
                params.append(key, query[key]);
            }
        }
        const string = params.toString();
        if (string) {
            if (url.includes('?')) {
                url += '&' + string;
            }
            else {
                url += '?' + string;
            }
        }
    }
    return url;
}
exports.addQuery = addQuery;
function anyToCamelCase(object, skip) {
    if (object === null) {
        return object;
    }
    if (typeof (object) === 'object') {
        if (Array.isArray(object)) {
            const obj = [];
            for (let value of object) {
                obj.push(anyToCamelCase(value));
            }
            return obj;
        }
        else {
            const obj = {};
            for (let key in object) {
                if (skip && skip.includes(key)) {
                    obj[key] = object[key];
                }
                else {
                    obj[toCamelCase(key)] = anyToCamelCase(object[key]);
                }
            }
            return obj;
        }
    }
    return object;
}
exports.anyToCamelCase = anyToCamelCase;
function getAcronym(name) {
    if (name != null) {
        return name.replace(/\w+/g, match => match[0]).replace(/\s/g, '');
    }
    return '';
}
exports.getAcronym = getAcronym;
function getExceededRatelimits(ratelimits, message, now = Date.now()) {
    const exceeded = [];
    for (const ratelimit of ratelimits) {
        let cacheId;
        switch (ratelimit.type) {
            case constants_1.CommandRatelimitTypes.CHANNEL:
                {
                    cacheId = message.channelId;
                }
                ;
                break;
            case constants_1.CommandRatelimitTypes.GUILD:
                {
                    cacheId = message.guildId || message.channelId;
                }
                ;
                break;
            default:
                {
                    cacheId = message.author.id;
                }
                ;
        }
        const item = ratelimit.get(cacheId);
        if (ratelimit.limit <= item.usages++) {
            const remaining = (item.start + ratelimit.duration) - now;
            exceeded.push({ item, ratelimit, remaining });
        }
    }
    return exceeded;
}
exports.getExceededRatelimits = getExceededRatelimits;
function getFormatFromHash(hash, format, defaultFormat = constants_1.ImageFormats.PNG) {
    if (format) {
        format = format.toLowerCase();
    }
    else {
        format = defaultFormat;
        if (hash.startsWith('a_')) {
            format = constants_1.ImageFormats.GIF;
        }
    }
    if (!constants_1.IMAGE_FORMATS.includes(format)) {
        throw new Error(`Invalid format: '${format}', valid: ${constants_1.IMAGE_FORMATS}`);
    }
    return format;
}
exports.getFormatFromHash = getFormatFromHash;
function hexToInt(hex) {
    return parseInt(hex.replace(/#/, ''), 16);
}
exports.hexToInt = hexToInt;
function intToHex(int, hashtag) {
    return ((hashtag) ? '#' : '') + int.toString(16).padStart(6, '0');
}
exports.intToHex = intToHex;
function intToRGB(int) {
    return {
        r: (int >> 16) & 0x0ff,
        g: (int >> 8) & 0x0ff,
        b: int & 0x0ff,
    };
}
exports.intToRGB = intToRGB;
function regex(type, content, onlyFirst = false) {
    type = String(type || '').toUpperCase();
    const regex = constants_1.DiscordRegex[type];
    if (regex === undefined) {
        throw new Error(`Unknown regex type: ${type}`);
    }
    regex.lastIndex = 0;
    const payload = {
        match: { regex, type },
        matches: [],
    };
    let match = null;
    while (match = regex.exec(content)) {
        const result = { matched: match[0] };
        switch (type) {
            case constants_1.DiscordRegexNames.EMOJI:
                {
                    result.name = match[1];
                    result.id = match[2];
                    result.animated = content.startsWith('<a:');
                }
                ;
                break;
            case constants_1.DiscordRegexNames.MENTION_CHANNEL:
            case constants_1.DiscordRegexNames.MENTION_ROLE:
                {
                    result.id = match[1];
                }
                ;
                break;
            case constants_1.DiscordRegexNames.MENTION_USER:
                {
                    result.id = match[2];
                    result.mentionType = match[1];
                }
                ;
                break;
            case constants_1.DiscordRegexNames.TEXT_CODEBLOCK:
                {
                    result.language = match[2];
                    result.text = match[3];
                }
                ;
                break;
            case constants_1.DiscordRegexNames.TEXT_BOLD:
            case constants_1.DiscordRegexNames.TEXT_CODESTRING:
            case constants_1.DiscordRegexNames.TEXT_ITALICS:
            case constants_1.DiscordRegexNames.TEXT_SNOWFLAKE:
            case constants_1.DiscordRegexNames.TEXT_SPOILER:
            case constants_1.DiscordRegexNames.TEXT_STRIKE:
            case constants_1.DiscordRegexNames.TEXT_UNDERLINE:
            case constants_1.DiscordRegexNames.TEXT_URL:
                {
                    result.text = match[1];
                }
                ;
                break;
            default:
                {
                    throw new Error(`Unknown regex type: ${type}`);
                }
                ;
        }
        payload.matches.push(result);
        if (onlyFirst) {
            break;
        }
    }
    regex.lastIndex = 0;
    return payload;
}
exports.regex = regex;
function rgbToInt(r, g, b) {
    return ((r & 0x0ff) << 16) | ((g & 0x0ff) << 8) | (b & 0x0ff);
}
exports.rgbToInt = rgbToInt;
function toCamelCase(value) {
    if (!value.includes('_')) {
        return value;
    }
    value = value
        .split('_')
        .map((v) => v.charAt(0).toUpperCase() + v.slice(1).toLowerCase())
        .join('');
    return value.charAt(0).toLowerCase() + value.slice(1);
}
exports.toCamelCase = toCamelCase;
