// detritus-client/lib/structures/basestructure.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("util");
const collections_1 = require("../collections");
const constants_1 = require("../constants");
const utils_1 = require("../utils");
function convertKey(snake) {
    if (snake in constants_1.DetritusKeys) {
        return constants_1.DetritusKeys[snake];
    }
    return utils_1.toCamelCase(snake);
}
exports.convertKey = convertKey;
/**
 * The most basic Structure class, every structure extends this
 * @category Structure
 */
class Structure {
    _getFromSnake(key) {
        return this[convertKey(key)];
    }
    _setFromSnake(key, value) {
        return this[convertKey(key)] = value;
    }
    difference(key, value) {
        if (value !== undefined) {
            const camelKey = convertKey(key);
            const old = this[camelKey];
            if (old !== undefined) {
                if (old instanceof BaseStructure) {
                    let differences = old.differences(value);
                    if (differences) {
                        return [true, differences];
                    }
                }
                else if (old instanceof collections_1.BaseCollection) {
                    if (old.size !== value.length) {
                        return [true, old.clone()];
                    }
                    else if (old.size) {
                        return [true, old.clone()];
                    }
                }
                else if (old instanceof collections_1.BaseSet) {
                    if (old.size !== value.length) {
                        return [true, old.clone()];
                    }
                    else if (old.size) {
                        return [true, old.clone()];
                    }
                }
                else if (old instanceof Date) {
                    if (value) {
                        if (old.getTime() !== (new Date(value)).getTime()) {
                            return [true, old];
                        }
                    }
                    else {
                        return [true, old];
                    }
                }
                else if (Array.isArray(old)) {
                }
                else {
                    if (old !== value) {
                        return [true, old];
                    }
                }
            }
        }
        return [false, null];
    }
    differences(data) {
        let hasDifferences = false;
        const obj = {};
        for (let key in data) {
            if (this._keysSkipDifference && this._keysSkipDifference.has(key)) {
                continue;
            }
            const [hasDifference, difference] = this.difference(key, data[key]);
            if (hasDifference) {
                obj[convertKey(key)] = difference;
                hasDifferences = true;
            }
        }
        if (hasDifferences) {
            return obj;
        }
        return null;
    }
    merge(data) {
        if (this._keys) {
            if (this._keysMerge) {
                for (let key of this._keysMerge) {
                    if (this._keys.has(key)) {
                        this.mergeValue(key, data[key]);
                    }
                }
            }
            for (let key in data) {
                if (this._keysMerge && this._keysMerge.has(key)) {
                    continue;
                }
                if (this._keys.has(key)) {
                    let value = data[key];
                    if (value instanceof BaseStructure) {
                        this._setFromSnake(key, value);
                        continue;
                    }
                    this.mergeValue(key, value);
                }
            }
        }
    }
    mergeValue(key, value) {
        if (value !== undefined) {
            this._setFromSnake(key, value);
        }
    }
    toJSON() {
        const obj = {};
        if (this._keys) {
            for (let key of this._keys) {
                obj[key] = this._getFromSnake(key);
            }
        }
        return obj;
    }
    [util_1.inspect.custom]() {
        // https://github.com/abalabahaha/eris/blob/master/lib/structures/Base.js#L59
        const copy = new ({ [this.constructor.name]: class {
            } })[this.constructor.name]();
        if (this._keys) {
            for (let key of this._keys) {
                key = convertKey(key);
                copy[key] = this[key];
            }
        }
        return copy;
    }
}
exports.Structure = Structure;
/**
 * Basic Structure class with an added ShardClient attached to it
 * @category Structure
 */
class BaseStructure extends Structure {
    constructor(client) {
        super();
        this.client = client;
    }
    get shardId() {
        return this.client.shardId;
    }
}
exports.BaseStructure = BaseStructure;
