// detritus-client/lib/structures/profile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const basecollection_1 = require("../collections/basecollection");
const baseset_1 = require("../collections/baseset");
const constants_1 = require("../constants");
const basestructure_1 = require("./basestructure");
const connectedaccount_1 = require("./connectedaccount");
const user_1 = require("./user");
const keysProfile = new baseset_1.BaseSet([
    constants_1.DiscordKeys.CONNECTED_ACCOUNTS,
    constants_1.DiscordKeys.MUTUAL_GUILDS,
    constants_1.DiscordKeys.PREMIUM_GUILD_SINCE,
    constants_1.DiscordKeys.PREMIUM_SINCE,
    constants_1.DiscordKeys.USER,
]);
/**
 * User Profile Structure
 * only non-bots will ever see these
 * @category Structure
 */
class Profile extends basestructure_1.BaseStructure {
    constructor(client, data) {
        super(client);
        this._keys = keysProfile;
        this.connectedAccounts = new basecollection_1.BaseCollection();
        this.mutualGuilds = new basecollection_1.BaseCollection();
        this.nicks = new basecollection_1.BaseCollection();
        this.premiumGuildSince = null;
        this.premiumSince = null;
        this.merge(data);
    }
    mergeValue(key, value) {
        if (value !== undefined) {
            switch (key) {
                case constants_1.DiscordKeys.CONNECTED_ACCOUNTS:
                    {
                        this.connectedAccounts.clear();
                        for (let raw of value) {
                            const account = new connectedaccount_1.ConnectedAccount(this.client, raw);
                            this.connectedAccounts.set(account.key, account);
                        }
                    }
                    ;
                    return;
                case constants_1.DiscordKeys.MUTUAL_GUILDS:
                    {
                        this.mutualGuilds.clear();
                        for (let raw of value) {
                            if (this.client.guilds.has(raw.id)) {
                                const guild = this.client.guilds.get(raw.id);
                                this.mutualGuilds.set(guild.id, guild);
                            }
                            else {
                                this.mutualGuilds.set(raw.id, null);
                            }
                        }
                    }
                    ;
                    return;
                case constants_1.DiscordKeys.PREMIUM_GUILD_SINCE:
                    {
                        if (value) {
                            value = new Date(value);
                        }
                    }
                    ;
                    break;
                case constants_1.DiscordKeys.PREMIUM_SINCE:
                    {
                        if (value) {
                            value = new Date(value);
                        }
                    }
                    ;
                    break;
                case constants_1.DiscordKeys.USER:
                    {
                        let user;
                        if (this.client.users.has(value.id)) {
                            user = this.client.users.get(value.id);
                            user.merge(value);
                        }
                        else {
                            user = new user_1.User(this.client, value);
                        }
                        value = user;
                    }
                    ;
                    break;
            }
            return super.mergeValue.call(this, key, value);
        }
    }
}
exports.Profile = Profile;
