// detritus-client/lib/structures/voicestate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const baseset_1 = require("../collections/baseset");
const constants_1 = require("../constants");
const basestructure_1 = require("./basestructure");
const member_1 = require("./member");
const keysVoiceState = new baseset_1.BaseSet([
    constants_1.DiscordKeys.CHANNEL_ID,
    constants_1.DiscordKeys.DEAF,
    constants_1.DiscordKeys.GUILD_ID,
    constants_1.DiscordKeys.MEMBER,
    constants_1.DiscordKeys.MUTE,
    constants_1.DiscordKeys.SELF_DEAF,
    constants_1.DiscordKeys.SELF_MUTE,
    constants_1.DiscordKeys.SELF_STREAM,
    constants_1.DiscordKeys.SELF_VIDEO,
    constants_1.DiscordKeys.SESSION_ID,
    constants_1.DiscordKeys.SUPPRESS,
    constants_1.DiscordKeys.USER_ID,
]);
const keysMergeVoiceState = new baseset_1.BaseSet([
    constants_1.DiscordKeys.GUILD_ID,
]);
const keysSkipDifferenceVoiceState = new baseset_1.BaseSet([
    constants_1.DiscordKeys.MEMBER,
]);
/**
 * Voice State Structure
 * @category Structure
 */
class VoiceState extends basestructure_1.BaseStructure {
    constructor(client, data) {
        super(client);
        this._keys = keysVoiceState;
        this._keysMerge = keysMergeVoiceState;
        this._keysSkipDifference = keysSkipDifferenceVoiceState;
        this.deaf = false;
        this.selfDeaf = false;
        this.selfMute = false;
        this.selfStream = false;
        this.selfVideo = false;
        this.sessionId = '';
        this.suppress = false;
        this.userId = '';
        this.merge(data);
    }
    get channel() {
        if (this.channelId) {
            return this.client.channels.get(this.channelId) || null;
        }
        return null;
    }
    get guild() {
        if (this.guildId) {
            return this.client.guilds.get(this.guildId) || null;
        }
        return null;
    }
    get serverId() {
        return this.guildId || this.channelId || '';
    }
    async edit(options) {
        if (!this.guildId) {
            throw new Error('Cannot edit a user in a DM call.');
        }
        return this.client.rest.editGuildMember(this.guildId, this.userId, options);
    }
    joinVoice(options) {
        return this.client.voiceConnect(this.guildId, this.channelId, options);
    }
    move(channelId) {
        return this.edit({ channelId });
    }
    setDeaf(deaf) {
        return this.edit({ deaf });
    }
    setMute(mute) {
        return this.edit({ mute });
    }
    mergeValue(key, value) {
        if (value !== undefined) {
            switch (key) {
                case constants_1.DiscordKeys.MEMBER:
                    {
                        const guildId = this.guildId;
                        let member;
                        if (this.client.members.has(guildId, value.user.id)) {
                            member = this.client.members.get(guildId, value.user.id);
                            member.merge(value);
                        }
                        else {
                            value.guild_id = guildId;
                            member = new member_1.Member(this.client, value);
                            this.client.members.insert(member);
                        }
                        value = member;
                    }
                    ;
                    break;
            }
            super.mergeValue.call(this, key, value);
        }
    }
}
exports.VoiceState = VoiceState;
