// detritus-client-rest/lib/clientsidechecks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Types = Object.freeze({
    ARRAY: 'array',
    BOOLEAN: 'boolean',
    NUMBER: 'number',
    OBJECT: 'object',
    SNOWFLAKE: 'snowflake',
    STRING: 'string',
});
exports.Regexes = {
    [exports.Types.SNOWFLAKE]: /^\d+|@me$/,
};
function bufferToBase64(buffer) {
    if (buffer instanceof Buffer) {
        return buffer.toString('base64');
    }
    return buffer;
}
exports.bufferToBase64 = bufferToBase64;
function verifyData(data, verification) {
    const verified = {};
    for (let key in verification) {
        const valueOptions = verification[key];
        if (!(key in data) || data[key] === undefined) {
            if (valueOptions.required) {
                throw new Error(`${key} is required.`);
            }
            continue;
        }
        let value = data[key];
        switch (valueOptions.type) {
            case exports.Types.ARRAY:
                {
                    if (!Array.isArray(value)) {
                        throw new Error(`${key} has to be an array.`);
                    }
                }
                ;
                break;
            case exports.Types.BOOLEAN:
                {
                    value = Boolean(value);
                }
                ;
                break;
            case exports.Types.NUMBER:
                {
                    value = parseInt(value);
                    if (value === NaN) {
                        throw new Error(`${key} has to be an integer.`);
                    }
                }
                ;
                break;
            case exports.Types.OBJECT:
                {
                    if (typeof (value) !== exports.Types.OBJECT) {
                        throw new Error(`${key} has to be an object.`);
                    }
                }
                ;
                break;
            case exports.Types.SNOWFLAKE:
                {
                    if (typeof (value) !== exports.Types.STRING && typeof (value) !== exports.Types.NUMBER) {
                        if (!exports.Regexes[exports.Types.SNOWFLAKE].exec(value)) {
                            throw new Error(`${key} has to be a snowflake.`);
                        }
                    }
                }
                ;
                break;
            case exports.Types.STRING:
                {
                    value = String(value);
                }
                ;
                break;
        }
        data[key] = value;
    }
}
exports.verifyData = verifyData;
