// detritus-rest/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Package = Object.freeze({
    URL: 'https://github.com/detritusjs/rest',
    VERSION: '0.3.1',
});
var HTTPHeaders;
(function (HTTPHeaders) {
    HTTPHeaders["HTTP2_AUTHORITY"] = ":authority";
    HTTPHeaders["HTTP2_METHOD"] = ":method";
    HTTPHeaders["HTTP2_PATH"] = ":path";
    HTTPHeaders["HTTP2_PROTOCOL"] = ":protocol";
    HTTPHeaders["HTTP2_SCHEME"] = ":scheme";
    HTTPHeaders["HTTP2_STATUS"] = ":status";
    HTTPHeaders["ACCEPT_ENCODING"] = "accept-encoding";
    HTTPHeaders["CONTENT_ENCODING"] = "content-encoding";
    HTTPHeaders["CONTENT_LENGTH"] = "content-length";
    HTTPHeaders["CONTENT_TYPE"] = "content-type";
    HTTPHeaders["USER_AGENT"] = "user-agent";
})(HTTPHeaders = exports.HTTPHeaders || (exports.HTTPHeaders = {}));
;
;
var HTTPMethods;
(function (HTTPMethods) {
    HTTPMethods["DELETE"] = "DELETE";
    HTTPMethods["GET"] = "GET";
    HTTPMethods["HEAD"] = "HEAD";
    HTTPMethods["OPTIONS"] = "OPTIONS";
    HTTPMethods["PATCH"] = "PATCH";
    HTTPMethods["POST"] = "POST";
    HTTPMethods["PUT"] = "PUT";
})(HTTPMethods = exports.HTTPMethods || (exports.HTTPMethods = {}));
;
exports.BodylessMethods = [
    HTTPMethods.GET,
    HTTPMethods.HEAD,
    HTTPMethods.OPTIONS,
];
var HTTPProtocols;
(function (HTTPProtocols) {
    HTTPProtocols["HTTP"] = "http:";
    HTTPProtocols["HTTPS"] = "https:";
})(HTTPProtocols = exports.HTTPProtocols || (exports.HTTPProtocols = {}));
;
var HTTPStatusCodes;
(function (HTTPStatusCodes) {
    HTTPStatusCodes[HTTPStatusCodes["CONTINUE"] = 100] = "CONTINUE";
    HTTPStatusCodes[HTTPStatusCodes["SWITCHING_PROTOCOLS"] = 101] = "SWITCHING_PROTOCOLS";
    HTTPStatusCodes[HTTPStatusCodes["PROCESSING"] = 102] = "PROCESSING";
    HTTPStatusCodes[HTTPStatusCodes["EARLY_HINTS"] = 103] = "EARLY_HINTS";
    HTTPStatusCodes[HTTPStatusCodes["OK"] = 200] = "OK";
    HTTPStatusCodes[HTTPStatusCodes["CREATED"] = 201] = "CREATED";
    HTTPStatusCodes[HTTPStatusCodes["ACCEPTED"] = 202] = "ACCEPTED";
    HTTPStatusCodes[HTTPStatusCodes["NON_AUTHORITATIVE_INFORMATION"] = 203] = "NON_AUTHORITATIVE_INFORMATION";
    HTTPStatusCodes[HTTPStatusCodes["NO_CONTENT"] = 204] = "NO_CONTENT";
    HTTPStatusCodes[HTTPStatusCodes["RESET_CONTENT"] = 205] = "RESET_CONTENT";
    HTTPStatusCodes[HTTPStatusCodes["PARTIAL_CONTENT"] = 206] = "PARTIAL_CONTENT";
    HTTPStatusCodes[HTTPStatusCodes["MULTI_STATUS"] = 207] = "MULTI_STATUS";
    HTTPStatusCodes[HTTPStatusCodes["ALREADY_REPORTED"] = 208] = "ALREADY_REPORTED";
    HTTPStatusCodes[HTTPStatusCodes["IM_USED"] = 226] = "IM_USED";
    HTTPStatusCodes[HTTPStatusCodes["MULTIPLE_CHOICES"] = 300] = "MULTIPLE_CHOICES";
    HTTPStatusCodes[HTTPStatusCodes["MOVED_PERMANENTLY"] = 301] = "MOVED_PERMANENTLY";
    HTTPStatusCodes[HTTPStatusCodes["FOUND"] = 302] = "FOUND";
    HTTPStatusCodes[HTTPStatusCodes["SEE_OTHER"] = 303] = "SEE_OTHER";
    HTTPStatusCodes[HTTPStatusCodes["NOT_MODIFIED"] = 304] = "NOT_MODIFIED";
    HTTPStatusCodes[HTTPStatusCodes["USE_PROXY"] = 305] = "USE_PROXY";
    HTTPStatusCodes[HTTPStatusCodes["TEMPORARY_REDIRECT"] = 307] = "TEMPORARY_REDIRECT";
    HTTPStatusCodes[HTTPStatusCodes["PERMANENT_REDIRECT"] = 308] = "PERMANENT_REDIRECT";
    HTTPStatusCodes[HTTPStatusCodes["BAD_REQUEST"] = 400] = "BAD_REQUEST";
    HTTPStatusCodes[HTTPStatusCodes["UNAUTHORIZED"] = 401] = "UNAUTHORIZED";
    HTTPStatusCodes[HTTPStatusCodes["PAYMENT_REQUIRED"] = 402] = "PAYMENT_REQUIRED";
    HTTPStatusCodes[HTTPStatusCodes["FORBIDDEN"] = 403] = "FORBIDDEN";
    HTTPStatusCodes[HTTPStatusCodes["NOT_FOUND"] = 404] = "NOT_FOUND";
    HTTPStatusCodes[HTTPStatusCodes["METHOD_NOT_ALLOWED"] = 405] = "METHOD_NOT_ALLOWED";
    HTTPStatusCodes[HTTPStatusCodes["NOT_ACCEPTABLE"] = 406] = "NOT_ACCEPTABLE";
    HTTPStatusCodes[HTTPStatusCodes["PROXY_AUTHENTICATION_REQUIRED"] = 407] = "PROXY_AUTHENTICATION_REQUIRED";
    HTTPStatusCodes[HTTPStatusCodes["REQUEST_TIMEOUT"] = 408] = "REQUEST_TIMEOUT";
    HTTPStatusCodes[HTTPStatusCodes["CONFLICT"] = 409] = "CONFLICT";
    HTTPStatusCodes[HTTPStatusCodes["GONE"] = 410] = "GONE";
    HTTPStatusCodes[HTTPStatusCodes["LENGTH_REQUIRED"] = 411] = "LENGTH_REQUIRED";
    HTTPStatusCodes[HTTPStatusCodes["PRECONDITION_FAILED"] = 412] = "PRECONDITION_FAILED";
    HTTPStatusCodes[HTTPStatusCodes["PAYLOAD_TOO_LARGE"] = 413] = "PAYLOAD_TOO_LARGE";
    HTTPStatusCodes[HTTPStatusCodes["URI_TOO_LONG"] = 414] = "URI_TOO_LONG";
    HTTPStatusCodes[HTTPStatusCodes["UNSUPPORTED_MEDIA_TYPE"] = 415] = "UNSUPPORTED_MEDIA_TYPE";
    HTTPStatusCodes[HTTPStatusCodes["RANGE_NOT_SATISFIABLE"] = 416] = "RANGE_NOT_SATISFIABLE";
    HTTPStatusCodes[HTTPStatusCodes["EXPECTATION_FAILED"] = 417] = "EXPECTATION_FAILED";
    HTTPStatusCodes[HTTPStatusCodes["TEAPOT"] = 418] = "TEAPOT";
    HTTPStatusCodes[HTTPStatusCodes["MISDIRECTED_REQUEST"] = 421] = "MISDIRECTED_REQUEST";
    HTTPStatusCodes[HTTPStatusCodes["UNPROCESSABLE_ENTITY"] = 422] = "UNPROCESSABLE_ENTITY";
    HTTPStatusCodes[HTTPStatusCodes["LOCKED"] = 423] = "LOCKED";
    HTTPStatusCodes[HTTPStatusCodes["FAILED_DEPENDENCY"] = 424] = "FAILED_DEPENDENCY";
    HTTPStatusCodes[HTTPStatusCodes["UNORDERED_COLLECTION"] = 425] = "UNORDERED_COLLECTION";
    HTTPStatusCodes[HTTPStatusCodes["UPGRADE_REQUIRED"] = 426] = "UPGRADE_REQUIRED";
    HTTPStatusCodes[HTTPStatusCodes["PRECONDITION_REQUIRED"] = 428] = "PRECONDITION_REQUIRED";
    HTTPStatusCodes[HTTPStatusCodes["TOO_MANY_REQUESTS"] = 429] = "TOO_MANY_REQUESTS";
    HTTPStatusCodes[HTTPStatusCodes["REQUEST_HEADER_FIELDS_TOO_LARGE"] = 431] = "REQUEST_HEADER_FIELDS_TOO_LARGE";
    HTTPStatusCodes[HTTPStatusCodes["UNAVAILABLE_FOR_LEGAL_REASONS"] = 451] = "UNAVAILABLE_FOR_LEGAL_REASONS";
    HTTPStatusCodes[HTTPStatusCodes["INTERNAL_SERVER_ERROR"] = 500] = "INTERNAL_SERVER_ERROR";
    HTTPStatusCodes[HTTPStatusCodes["NOT_IMPLEMENTED"] = 501] = "NOT_IMPLEMENTED";
    HTTPStatusCodes[HTTPStatusCodes["BAD_GATEWAY"] = 502] = "BAD_GATEWAY";
    HTTPStatusCodes[HTTPStatusCodes["SERVICE_UNAVAILABLE"] = 503] = "SERVICE_UNAVAILABLE";
    HTTPStatusCodes[HTTPStatusCodes["GATEWAY_TIMEOUT"] = 504] = "GATEWAY_TIMEOUT";
    HTTPStatusCodes[HTTPStatusCodes["HTTP_VERSION_NOT_SUPPORTED"] = 505] = "HTTP_VERSION_NOT_SUPPORTED";
    HTTPStatusCodes[HTTPStatusCodes["VARIANT_ALSO_NEGOTIATES"] = 506] = "VARIANT_ALSO_NEGOTIATES";
    HTTPStatusCodes[HTTPStatusCodes["INSUFFICIENT_STORAGE"] = 507] = "INSUFFICIENT_STORAGE";
    HTTPStatusCodes[HTTPStatusCodes["LOOP_DETECTED"] = 508] = "LOOP_DETECTED";
    HTTPStatusCodes[HTTPStatusCodes["BANDWIDTH_LIMIT_EXCEEDED"] = 509] = "BANDWIDTH_LIMIT_EXCEEDED";
    HTTPStatusCodes[HTTPStatusCodes["NOT_EXTENDED"] = 510] = "NOT_EXTENDED";
    HTTPStatusCodes[HTTPStatusCodes["NETWORK_AUTHENTICATION_REQUIRED"] = 511] = "NETWORK_AUTHENTICATION_REQUIRED";
})(HTTPStatusCodes = exports.HTTPStatusCodes || (exports.HTTPStatusCodes = {}));
;
exports.HTTPSPort = 443;
var ALPNProtocols;
(function (ALPNProtocols) {
    ALPNProtocols["NONE"] = "NONE";
    ALPNProtocols["HTTP1"] = "http/1.0";
    ALPNProtocols["HTTP1_1"] = "http/1.1";
    ALPNProtocols["HTTP2"] = "h2";
})(ALPNProtocols = exports.ALPNProtocols || (exports.ALPNProtocols = {}));
;
var SupportedContentTypes;
(function (SupportedContentTypes) {
    SupportedContentTypes["APPLICATION_JSON"] = "application/json";
    SupportedContentTypes["APPLICATION_OCTET_STREAM"] = "application/octet-stream";
    SupportedContentTypes["MULTIPART_FORM_DATA"] = "multipart/form-data";
    SupportedContentTypes["TEXT_PLAIN"] = "text/plain";
})(SupportedContentTypes = exports.SupportedContentTypes || (exports.SupportedContentTypes = {}));
;
